---
title: "CreateTally"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CreateTally}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup, include = FALSE}
library(brentlabRnaSeqTools)
```

```{r}
# NOTE: api_url is an environmental variable of the package that stores the current url to the crypto database api
combined_df = getMetadata(api_url)
```

```{r}
# grant df is a dataframe of the 2016 grant summary, loaded with the package
induction_set = createNinetyMinuteInductionSet(combined_df, grant_df)
passing_qc1_induction_set = qualityAssessmentFilter(induction_set)
```
### read in a deseq data set for this data that has had the DESeq function called on it (this should be done on the cluster)
```{r}
# this is in my scratch space (until the garbage collector gets it) at /scratch/mblab/chasem/rnaseq_pipeline/experiments/database_90_min_update. Feel free to rsync or scp it.
path = "/home/chase/projects/brentlab/daily_work/20210211_induction_set_tally_with_database/database_pulled_libraryptocol_librarydate_20210227.rds"
deseq_object = readRDS(path)
```
### extract norm counts
```{r}
norm_counts = counts(deseq_object, normalized=TRUE)
```
### remove the library protocol and date effects
```{r}
# NOTE: in this case, since there are no "wanted" covariates, we pass the entire design matrix (library protocol + librarydate).
# If genotype was included, then we would need to subset the design matrix to just the nuisance variable columns
removed_librarydate_effect_log_norm_counts = removeParameterEffects(deseq_object, design(deseq_object))
```
### RLE
```{r}
# obviously, change the output paths
norm_output = "/home/chase/projects/brentlab/daily_work/20210211_induction_set_tally_with_database/rle/norm_counts_rle"
removed_effect_output = "/home/chase/projects/brentlab/daily_work/20210211_induction_set_tally_with_database/rle/librarydate_effect_removed"
dir.create(norm_output)
dir.create(removed_effect_output)

sorted_passing_induction_meta_qual = as_tibble(colData(deseq_object))
```

```{r}
sorted_passing_induction_meta_qual$GENOTYPE = sorted_passing_induction_meta_qual$GENOTYPE1
extractRLEByReplicateGroup_90minInduction(sorted_passing_induction_meta_qual, norm_counts, norm_output, protocol_selector = FALSE, already_logged_flag = FALSE)
createRLEPlotsByReplicateGroup(norm_output, sorted_passing_induction_meta_qual)
```

```{r}
extractRLEByReplicateGroup_90minInduction(sorted_passing_induction_meta_qual, removed_librarydate_effect_log_norm_counts, removed_effect_output, protocol_selector = FALSE, already_logged_flag = TRUE)
createRLEPlotsByReplicateGroup(removed_effect_output, sorted_passing_induction_meta_qual)
```

```{r}

norm_counts_rle_summary_list = Sys.glob(paste(norm_output,"*summary*", sep="/"))
norm_counts_rle_summary_df_list = lapply(norm_counts_rle_summary_list, read_csv)
norm_counts_rle_summary = bind_rows(norm_counts_rle_summary_df_list)

norm_counts_rle_summary = left_join(norm_counts_rle_summary, sorted_passing_induction_meta_qual, on="FASTQFILENAME")

removed_effect_rle_summary_list = Sys.glob(paste(removed_effect_output,"*summary*", sep="/"))
removed_effect_rle_summary_df_list = lapply(removed_effect_rle_summary_list, read_csv)

removed_effect_rle_summary = bind_rows(removed_effect_rle_summary_df_list)

removed_effect_rle_summary = left_join(removed_effect_rle_summary, sorted_passing_induction_meta_qual, on="FASTQFILENAME")

```
### filter samples
```{r}
# TODO: UPDATE RLE PLOTS

iqr_fltr_rle_summary = removed_effect_rle_summary %>% filter(INTERQUARTILE_RANGE<.6125)


# write_csv(iqr_fltr_rle_summary, 'iqr_fltr_meta_qual.csv')
```

```{r}
plotRLE_bar_iqr = function(rle_summary_df, title){
  x =ggplot(rle_summary_df, aes(INTERQUARTILE_RANGE))+
    geom_histogram()+ggtitle(title)
  # dev.off()
  return(x)
}

hist = plotRLE_bar_iqr(removed_effect_rle_summary, "iqr")

plot(hist)
```

### create tally
```{r}
induction_set_tally = createInductionSetTally(induction_set, sorted_passing_induction_meta_qual, iqr_fltr_rle_summary, grant_df)
write_csv(induction_set_tally, "~/code/brentlab/induction_experiment_table/data/genotype_tally.csv")
```
### parse replicate sets out for the website table
```{r}
induction_meta_qual_with_iqr = induction_set %>% left_join(removed_effect_rle_summary %>% select(FASTQFILENAME, INTERQUARTILE_RANGE, ABS_SAMPLE_DEVIATION_MEDIAN))
# DO THIS FOR VISUALIZATION TO FIND BROWSER SHOT
induction_meta_qual_with_iqr$FASTQFILENAME = paste0(induction_meta_qual_with_iqr$FASTQFILENAME, '.fastq.gz')

table_vis_meta_qual = induction_meta_qual_with_iqr %>% select(AUTOAUDIT, AUTOSTATUSDECOMP, MANUALAUDIT, MANUALSTATUS, FASTQFILENAME, RUNNUMBER, LIBRARYDATE, GENOTYPE1, STRAIN, PROTEINCODINGTOTAL, NOTALIGNEDTOTALPERCENT, INTERQUARTILE_RANGE, ABS_SAMPLE_DEVIATION_MEDIAN, GENOTYPE1COVERAGE, GENOTYPE1LOG2CPM, NATCOVERAGE, NATLOG2CPM, G418COVERAGE, G418LOG2CPM)
write_csv(table_vis_meta_qual, '~/code/brentlab/induction_experiment_table/data/sample_detail_meta_qual.csv')
```

```{r}
perturbed_induction_set = table_vis_meta_qual %>% filter(GENOTYPE1 != 'CNAG_00000')

induction_set_split = split(perturbed_induction_set, f=list(perturbed_induction_set$GENOTYPE1))
```

```{r}
dir.create("~/code/brentlab/induction_experiment_table/data/split_by_genotype")
for(geno in names(induction_set_split)){
  print(geno)
  df = induction_set_split[[geno]]
  filename = str_trim(paste0(geno, ".csv"),side="both")
  write_csv(df, paste0('~/code/brentlab/induction_experiment_table/data/split_by_genotype/', filename))
}
```










