---
title: "database_interaction"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{database_interaction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup, include=TRUE}
library(brentlabRnaSeqTools)
library(RPostgres)
```


Below is an example of updating a record. Note that the slots of databse_info have been changed to non-functioning values. Check database_info itself for database options (you can do this by typing in database_info and tab or ?database_info). Please do not 'test' this code, or any of the code below, on a live database. However, you could create a playground sqlite database in memory and test these commands out there.

NOTE: there are syntax differences in the queries between sqlite and postgresql. The most significant is that postgresql is very picky about formating/capitalization and requires quotes around tables and maybe field names as a result. Check the postgresql documentation.
```{r update a record}
kn99_db = connectToDatabase(database_info$database_host, database_info$database_name, Sys.getenv('DB_USERNAME'), Sys.getenv('DB_PASSWORD'))

sql_stmt = glue_sql('UPDATE "bioSample"
                    SET genotype1 = \'CNAG_00830\'
                    WHERE "bioSample"."bioSampleNumber"
                      IN (2875, 2876)')
dbSendQuery(kn99_db, sql_stmt)

```

```{r connect to the database}
kn99_database = dbConnect(RPostgres::Postgres(),dbname = database_info$kn99_db_name,
                 host = database_info$kn99_host, # i.e. 'ec2-54-83-201-96.compute-1.amazonaws.com'
                 port = 5432, # or any other port specified by your DBA
                 user = Sys.getenv("db_user"),
                 password = Sys.getenv("db_password"))
```

Right now, there is a fair amount of manipulating that needs to happen to the fastq sheet -- do check that the columns and entries are what are expected. However, the error messages from the database are fairly useful

```{r read in fastq}
fastq_path="/path/to_new/fastq_sheets_to_database/fastq_05.08.21.xlsx"

new_sheet = normalizeFastqSheet(fastq_path)
```


```{r submit to database}

dbWriteTable(kn99_database, "fastqFiles", new_sheet, append=TRUE)

```
