---
title: "Yeast ZEV experiments"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    social: menu
    source_code: embed
runtime: shiny
---

```{r setup, include = FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(brentlabRnaSeqTools)
library(scales)
library(gprofiler2)
```

```{r gather data from cluster, include=FALSE}
# you can use this codeblock to read in the raw counts by reading in all of the raw count files from your align_count directory. Note: if you have both yeast and crypto raw counts, there will be some issues you'll need to address here. If you have only one organism in your align_counts_results dir, this works directly

# get counts by collating the run count output
# username = "chasem"
# path_list = Sys.glob(paste0("/mnt/htcf_scratch/",username, "/rnaseq_pipeline/align_count_results/*/*raw*"))
# df_list = lapply(path_list, read_csv)
# # the following is in genome_files -- it is in your rnaseq_pipeline directory
gene_id_path = '/home/chase/Desktop/rnaseq_pipeline/rnaseq_pipeline/genome_files/S288C_R64/S288C_R64_gids'
yeast_gene_id_column = read_csv(gene_id_path, col_names = "gene_id")
# df_list = lapply(df_list, function(x) select(x, -gene_id))
# 
# yeast_counts = bind_cols(df_list)
# 
# # get metadata from the sheet used to create the align_count_results output
# sample_metadata_csv = "yeast_20210423.csv"
# yeast_metadata = read_csv(paste0("/mnt/htcf_scratch/",username, "/rnaseq_pipeline/query/", sample_metadata_csv))
# 
# # get rid of crypto samples included by accident
# yeast_metadata = yeast_metadata %>% filter(!grepl("CKF44", genotype1))
# 
# # remove suffixes from the sample identifiers
# colnames(yeast_counts) = str_remove(colnames(yeast_counts), "_read_count.tsv")
# # remove suffix from fastqFileName col
# yeast_metadata$fastqFileName = str_remove(yeast_metadata$fastqFileName, ".fastq.gz")
# 
# # this is also technically unnecessary in this case, but included as an example of how to filter the counts down to just the samples in the metadata.
# yeast_counts = yeast_counts[, yeast_metadata$fastqFileName]
```

```{r verify col and row order, include=FALSE, echo=FALSE}
# check to ensure that the raw counts columns are in the same order as the metadata rows
# all.equal(colnames(yeast_counts), yeast_metadata$fastqFileName)

```

```{r run deseq, include=FALSE, echo=FALSE, cache=TRUE}
# dds = DESeqDataSetFromMatrix(colData = yeast_metadata, countData = yeast_counts, design=~timePoint+treatment)
# 
# # not necessary for this analysis -- could call estimateSizeFactors instead to get normalized counts -- but if you wanted to look at DE, you could
# deseq_obj = DESeq(dds)
# write_rds(deseq_obj, "../personal_eda/yeast/data/full_yeast_dataset_20210424.rds")

```
```{r load data from file, include=FALSE, echo=FALSE, cache=TRUE}
yeast_metadata = read_csv("../personal_eda/yeast/data/yeast_metadata_20210424.csv")
deseq_obj = readRDS("../personal_eda/yeast/data/full_yeast_deseq_20210424.rds")
```

```{r extract norm counts, include=FALSE, echo=FALSE, cache=TRUE}
norm_counts = counts(deseq_obj, normalized=TRUE)
rownames(norm_counts) = yeast_gene_id_column$gene_id
```

```{r translate gene id to gene name, include=FALSE, echo=FALSE}
# translate gene_id to gene_names and get other info
converted_gene_id = gconvert(yeast_gene_id_column$gene_id, 'scerevisiae', "ENSG", mthreshold=1, filter_na=TRUE)
# extract the genotypes of interest
fltr_converted_gene_id = converted_gene_id %>% filter(name %in% unique(yeast_metadata$genotype1)) %>% select(input, name, description)
```


```{r}
# again, note that the sample_norm_counts need rownames assigned to the gene_ids
# graphYeastTimeCourse(yeast_metadata, 'CST6', 'YIL036W', norm_counts)
```


```{r global, include=FALSE}
# load data in 'global' chunk so it can be shared by all users of the dashboard
library(datasets)
data(faithful)
```

Column {.sidebar}
-----------------------------------------------------------------------

Select a gene

```{r}
selectInput("gene_name", label = "Gene Name:",
            choices = unique(yeast_metadata$genotype1), selected = "CST6")
```

Column
-----------------------------------------------------------------------

### Graphs faceted on ruNumber and libraryDate

```{r}
renderPlot({
  gene_name = as.character(input$gene_name)
  gene_id = as.character(fltr_converted_gene_id[fltr_converted_gene_id$name == gene_name, 'input'])
  graphYeastTimeCourse(yeast_metadata, gene_name, gene_id, norm_counts)
})
```



<!-- # NOTE: the stuff below is miscellaneous code from an older iteration of this that I haven't gone through yet. I'm keeping it until I do in case there is something useful. -->


<!-- # in order CST6, RGM1, GAL4, RGT1, which is order of unique(query_df$genotype) -->
<!-- tf_systematic_names = c('YIL036W', 'YMR182C', 'YPL248C','YKL038W') -->
<!-- names(tf_systematic_names) = unique(query_df$genotype) -->

<!-- # create connected line graphs of each tf in tf_systematic_names -->
<!-- for (name in names(tf_systematic_names)){ -->
<!--   gene_symbol = name -->
<!--   systematic_name = tf_systematic_names[[name]] -->
<!--   g = norm_count_meta_df %>% -->
<!--     filter(genotype == gene_symbol, gene_id == systematic_name) %>% -->
<!--     ggplot(aes(timePoint, norm_count, color=treatment))+ -->
<!--     geom_line(aes(group=interaction(libraryDate, treatment, replicate)))+ -->
<!--     geom_point()+ -->
<!--     scale_y_continuous(trans = log2_trans(), -->
<!--                        breaks = trans_breaks("log2", function(x) 2^x), -->
<!--                        labels = trans_format("log2", math_format(2^.x)))+ -->
<!--     ggtitle(gene_symbol)+ -->
<!--     facet_wrap(~runNumber+libraryDate) -->
<!--     print(g) -->

<!-- } -->

<!-- # visualize number of DE genes at different thresholds -->
<!-- # do this to get all counts conditions -->
<!-- de_total_df = de_results_df %>% -->
<!--   group_by(library_conditions) %>% -->
<!--   count() %>% -->
<!--   rename(total = n)  -->

<!-- # filter -- change values of filter to examine different thresholds -->
<!-- de_change_df = de_results_df %>% -->
<!--   filter(padj < .05, log2FoldChange > .7) %>% -->
<!--   group_by(library_conditions) %>% -->
<!--   count() %>% -->
<!--   rename(filtered_total = n) -->

<!-- # merge -->
<!-- num_de_genes_df = left_join(de_total_df, de_change_df) -->
<!-- # set NA values in filtered_total (rows with no DE genes) to 0 -->
<!-- num_de_genes_df$filtered_total[is.na(df$filtered_total)] = 0 -->

<!-- # drop the total column and keep only the filtered_total -->
<!-- num_de_genes_df = num_de_genes_df %>% -->
<!--   select(filtered_total) -->

<!-- # add columns genotype, treatment, timepoint_contrast for graphing -->
<!-- num_de_genes_df$genotype = rep(c('CST6', 'GAL4', 'RGM1', 'RGT1'), each=8) -->
<!-- num_de_genes_df$treatment = rep(c('Estradiol', 'EtOH'), each = 4, times = 4) -->
<!-- num_de_genes_df$timepoint_contrast = rep(c('1_15', '1_20', '1_30', '1_90'), times = 8) -->

<!-- # force NA to 0 if there are no de genes -->
<!-- num_de_genes_df[is.na(num_de_genes_df)] = 0 -->

<!-- # visualize number of differentiallye expressed genes -->
<!-- num_de_genes_df %>% -->
<!--   ggplot(aes(timepoint_contrast, fill = treatment))+ -->
<!--     geom_bar(aes(weight=filtered_total), position='dodge')+ -->
<!--     scale_y_continuous(trans = log2_trans(), -->
<!--                      breaks = trans_breaks("log2", function(x) 2^x), -->
<!--                      labels = trans_format("log2", math_format(2^.x)))+ -->
<!--     facet_wrap(~genotype) -->




