---
title: "yeast_timecourse_qc"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{yeast_timecourse_qc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(brentlabRnaSeqTools)
library(scales)
```

# NOTE: the data needed to run this notebook verbatim comes packaged and ready to use. I include the codeblock below as an example of how to get other similar data from the cluster into R

```{r, include=FALSE}
# you can use this codeblock to read in the raw counts by reading in all of the raw count files from your align_count directory. Note: if you have both yeast and crypto raw counts, there will be some issues you'll need to address here. If you have only one organism in your align_counts_results dir, this works directly

# # get counts by collating the run count output
# username = "chasem"
# path_list = Sys.glob(paste0("/mnt/htcf_scratch/",username, "/rnaseq_pipeline/align_count_results/*/*raw*"))
# df_list = lapply(path_list, read_csv)
# # the following is in genome_files -- it is in your rnaseq_pipeline directory
# gene_id_path = '/home/chase/Desktop/rnaseq_pipeline/rnaseq_pipeline/genome_files/S288C_R64/S288C_R64_gids'
# yeast_gene_id_column = read_csv(gene_id_path, col_names = "gene_id")
# df_list = lapply(df_list, function(x) select(x, -gene_id))
# 
# cst6_raw_counts = bind_cols(df_list)
# 
# # get metadata from the sheet used to create the align_count_results output
# sample_metadata_csv = "cst6_sample_data.csv"
# cst6_sample_metadata = read_csv(paste0("/mnt/htcf_scratch/",username, "/rnaseq_pipeline/query/", sample_metadata_csv))

```

```{r}
# remove suffixes from the sample identifiers
colnames(cst6_raw_counts) = str_remove(colnames(cst6_raw_counts), "_read_count.tsv")
# remove suffix from fastqFileName col
cst6_sample_metadata$fastqFileName = str_remove(cst6_sample_metadata$fastqFileName, ".fastq.gz")
```


```{r}
# note: I added cst6 sample data as stored data in this module. If you want to examine different samples, you'd simply read in the metadata with read_csv

# this is also technically unnecessary in this case, but included as an example of how to filter the counts down to just the samples in the metadata.
cst6_raw_counts = cst6_raw_counts[, cst6_sample_metadata$fastqFileName]

# check to ensure that the raw counts columns are in the same order as the metadata rows
all.equal(colnames(cst6_raw_counts), cst6_sample_metadata$fastqFileName)

```

```{r}
dds = DESeqDataSetFromMatrix(colData = cst6_sample_metadata, countData = cst6_raw_counts, design=~timePoint+treatment)
```

```{r}
# not necessary for this analysis -- could call estimateSizeFactors instead to get normalized counts -- but if you wanted to look at DE, you could
deseq_obj = DESeq(dds)
```
```{r}
cst6_norm_counts = counts(deseq_obj, normalized=TRUE)
rownames(cst6_norm_counts) = yeast_gene_id_column$gene_id
```

```{r}

# again, note that the sample_norm_counts need rownames assigned to the gene_ids
graphYeastTimeCourse(cst6_sample_metadata, 'CST6', 'YIL036W', cst6_norm_counts)
```


# NOTE: the stuff below is miscellaneous code from an older iteration of this that I haven't gone through yet. I'm keeping it until I do in case there is something useful.

```{r}
# in order CST6, RGM1, GAL4, RGT1, which is order of unique(query_df$genotype)
tf_systematic_names = c('YIL036W', 'YMR182C', 'YPL248C','YKL038W')
names(tf_systematic_names) = unique(query_df$genotype)

# create connected line graphs of each tf in tf_systematic_names
for (name in names(tf_systematic_names)){
  gene_symbol = name
  systematic_name = tf_systematic_names[[name]]
  g = norm_count_meta_df %>%
    filter(genotype == gene_symbol, gene_id == systematic_name) %>%
    ggplot(aes(timePoint, norm_count, color=treatment))+
    geom_line(aes(group=interaction(libraryDate, treatment, replicate)))+
    geom_point()+
    scale_y_continuous(trans = log2_trans(),
                       breaks = trans_breaks("log2", function(x) 2^x),
                       labels = trans_format("log2", math_format(2^.x)))+
    ggtitle(gene_symbol)+
    facet_wrap(~runNumber+libraryDate)
    print(g)
  
}

# visualize number of DE genes at different thresholds
# do this to get all counts conditions
de_total_df = de_results_df %>%
  group_by(library_conditions) %>%
  count() %>%
  rename(total = n) 

# filter -- change values of filter to examine different thresholds
de_change_df = de_results_df %>%
  filter(padj < .05, log2FoldChange > .7) %>%
  group_by(library_conditions) %>%
  count() %>%
  rename(filtered_total = n)

# merge
num_de_genes_df = left_join(de_total_df, de_change_df)
# set NA values in filtered_total (rows with no DE genes) to 0
num_de_genes_df$filtered_total[is.na(df$filtered_total)] = 0

# drop the total column and keep only the filtered_total
num_de_genes_df = num_de_genes_df %>%
  select(filtered_total)

# add columns genotype, treatment, timepoint_contrast for graphing
num_de_genes_df$genotype = rep(c('CST6', 'GAL4', 'RGM1', 'RGT1'), each=8)
num_de_genes_df$treatment = rep(c('Estradiol', 'EtOH'), each = 4, times = 4)
num_de_genes_df$timepoint_contrast = rep(c('1_15', '1_20', '1_30', '1_90'), times = 8)

# force NA to 0 if there are no de genes
num_de_genes_df[is.na(num_de_genes_df)] = 0

# visualize number of differentiallye expressed genes
num_de_genes_df %>%
  ggplot(aes(timepoint_contrast, fill = treatment))+
    geom_bar(aes(weight=filtered_total), position='dodge')+
    scale_y_continuous(trans = log2_trans(),
                     breaks = trans_breaks("log2", function(x) 2^x),
                     labels = trans_format("log2", math_format(2^.x)))+
    facet_wrap(~genotype)
```



