<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Count reads in a library like HTSeq — countLibrary • brentlabRnaSeqTools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Count reads in a library like HTSeq — countLibrary" />
<meta property="og:description" content="This function creates a granges object from the txdb and calls countReadsInRanges.
Uses for this function may be during QC to generate log2cpm over the CDS regions in protein
coding transcriptome, for example" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">brentlabRnaSeqTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/create_nextflow_samplesheet.html">create_nextflow_samplesheet</a>
    </li>
    <li>
      <a href="../articles/create_set.html">create_set</a>
    </li>
    <li>
      <a href="../articles/database_interaction.html">database_interaction</a>
    </li>
    <li>
      <a href="../articles/experiment_sets.html">experiment_sets</a>
    </li>
    <li>
      <a href="../articles/qcExplorer.html">qcExplorer</a>
    </li>
    <li>
      <a href="../articles/running_the_nextflow_pipeline.html">running_the_nextflow_pipeline</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Count reads in a library like HTSeq</h1>
    <small class="dont-index">Source: <a href='https://github.com/cmatKhan/brentlabRnaSeqTools/blob/master/R/BamFunctions.R'><code>R/BamFunctions.R</code></a></small>
    <div class="hidden name"><code>countLibrary.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function creates a granges object from the txdb and calls <code><a href='countReadsInRanges.html'>countReadsInRanges</a></code>.
Uses for this function may be during QC to generate log2cpm over the CDS regions in protein
coding transcriptome, for example</p>
    </div>

    <pre class="usage"><span class='fu'>countLibrary</span><span class='op'>(</span>
  <span class='va'>bam_path</span>,
  <span class='va'>annote_db</span>,
  <span class='va'>feature_type</span>,
  <span class='va'>strandedness</span>,
  num_threads <span class='op'>=</span> <span class='fl'>1</span>,
  single_end_reads_flag <span class='op'>=</span> <span class='cn'>TRUE</span>,
  mapq_filter <span class='op'>=</span> <span class='fl'>20L</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bam_path</th>
      <td><p>path to a bam file. the index must also exist</p></td>
    </tr>
    <tr>
      <th>annote_db</th>
      <td><p>txdb object. see <code><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></code></p></td>
    </tr>
    <tr>
      <th>feature_type</th>
      <td><p>one of c("exon", "cds"). Determines what feature to count over</p></td>
    </tr>
    <tr>
      <th>strandedness</th>
      <td><p>strandedness of library determined by protocol. One of c("reverse", "same", "unstranded")
reverse means the counted reads will be on the opposite strand of the feature, same means they
will be on the same strand. unstranded counts reads over the feature regardless of strand</p></td>
    </tr>
    <tr>
      <th>num_threads</th>
      <td><p>number of threads available for parallelization. Default to 1</p></td>
    </tr>
    <tr>
      <th>single_end_reads_flag</th>
      <td><p>whether or not the library is single or paired. Note: this is not written to deal
with paired end reads currently, though you can try</p></td>
    </tr>
    <tr>
      <th>mapq_filter</th>
      <td><p>quality filter on read alignment. Default to 20, same as HTSeq.
see <a href='https://htseq.readthedocs.io/en/release_0.11.1/count.html'>https://htseq.readthedocs.io/en/release_0.11.1/count.html</a></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A RangedSummarizedExperiment. See <code><a href='https://rdrr.io/pkg/GenomicAlignments/man/summarizeOverlaps-methods.html'>summarizeOverlaps</a></code></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='http://achri.blogspot.com/2016/02/how-not-to-use-deseq2-for-illumina.html'>http://achri.blogspot.com/2016/02/how-not-to-use-deseq2-for-illumina.html</a>,
<a href='https://htseq.readthedocs.io/en/release_0.11.1/count.html'>https://htseq.readthedocs.io/en/release_0.11.1/count.html</a>,
<code><a href='https://rdrr.io/pkg/GenomicAlignments/man/summarizeOverlaps-methods.html'>summarizeOverlaps</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/cmatKhan/brentlabRnaSeqTools'>brentlabRnaSeqTools</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/AnnotationDbi'>AnnotationDbi</a></span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'BiocGenerics'</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:parallel':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:stats':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
#&gt; <span class='message'>    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
#&gt; <span class='message'>    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
#&gt; <span class='message'>    union, unique, unsplit, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'S4Vectors'</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid, I, unname</span></div><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>── <span style='font-weight: bold;'>Attaching packages</span> ─────────────────────────────────────── tidyverse 1.3.1 ──</span></div><div class='output co'>#&gt; <span class='message'><span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>ggplot2</span> 3.3.5     <span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>dplyr  </span> 1.0.7</span>
#&gt; <span class='message'><span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>tidyr  </span> 1.1.3     <span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>stringr</span> 1.4.0</span>
#&gt; <span class='message'><span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>readr  </span> 2.0.0     <span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>forcats</span> 0.5.1</span>
#&gt; <span class='message'><span style='color: #00BB00;'>✔</span> <span style='color: #0000BB;'>purrr  </span> 0.3.4     </span></div><div class='output co'>#&gt; <span class='message'>── <span style='font-weight: bold;'>Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>collapse()</span>   masks <span style='color: #0000BB;'>IRanges</span>::collapse()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>combine()</span>    masks <span style='color: #0000BB;'>Biobase</span>::combine(), <span style='color: #0000BB;'>BiocGenerics</span>::combine()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>desc()</span>       masks <span style='color: #0000BB;'>IRanges</span>::desc()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>tidyr</span>::<span style='color: #00BB00;'>expand()</span>     masks <span style='color: #0000BB;'>S4Vectors</span>::expand()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>filter()</span>     masks <span style='color: #0000BB;'>stats</span>::filter()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>first()</span>      masks <span style='color: #0000BB;'>S4Vectors</span>::first()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>lag()</span>        masks <span style='color: #0000BB;'>stats</span>::lag()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>ggplot2</span>::<span style='color: #00BB00;'>Position()</span> masks <span style='color: #0000BB;'>BiocGenerics</span>::Position(), <span style='color: #0000BB;'>base</span>::Position()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>purrr</span>::<span style='color: #00BB00;'>reduce()</span>     masks <span style='color: #0000BB;'>IRanges</span>::reduce()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>rename()</span>     masks <span style='color: #0000BB;'>S4Vectors</span>::rename()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>select()</span>     masks <span style='color: #0000BB;'>AnnotationDbi</span>::select()</span>
#&gt; <span class='message'><span style='color: #BB0000;'>✖</span> <span style='color: #0000BB;'>dplyr</span>::<span style='color: #00BB00;'>slice()</span>      masks <span style='color: #0000BB;'>IRanges</span>::slice()</span></div><div class='input'>
<span class='va'>kn99_db</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html'>loadDb</a></span><span class='op'>(</span><span class='st'>"data/kn99_db.sqlite"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in dbFileConnect(file): DB file 'data/kn99_db.sqlite' not found</span></div><div class='input'>
<span class='va'>combined_df</span> <span class='op'>=</span> <span class='fu'><a href='getMetadata.html'>getMetadata</a></span><span class='op'>(</span><span class='va'>database_info</span><span class='op'>$</span><span class='va'>kn99_host</span>,
                          <span class='va'>database_info</span><span class='op'>$</span><span class='va'>kn99_db_name</span>,
                          <span class='fu'><a href='https://rdrr.io/r/base/Sys.getenv.html'>Sys.getenv</a></span><span class='op'>(</span><span class='st'>"db_username"</span><span class='op'>)</span>,
                          <span class='fu'><a href='https://rdrr.io/r/base/Sys.getenv.html'>Sys.getenv</a></span><span class='op'>(</span><span class='st'>"password"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error: fe_sendauth: no password supplied</span></div><div class='input'>
<span class='va'>run_5102</span> <span class='op'>=</span> <span class='va'>combined_df</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>runNumber</span> <span class='op'>==</span> <span class='fl'>5102</span>,
        <span class='va'>experimentDesign</span> <span class='op'>==</span> <span class='st'>'90minuteInduction'</span>,
        <span class='va'>genotype1</span> <span class='op'>==</span> <span class='st'>"CNAG_00000"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in filter(., runNumber == 5102, experimentDesign == "90minuteInduction",     genotype1 == "CNAG_00000"): object 'combined_df' not found</span></div><div class='input'>
<span class='va'>wt_fastq</span> <span class='op'>=</span> <span class='va'>run_5102</span><span class='op'>[[</span><span class='fl'>1</span>, <span class='st'>'fastqFileName'</span><span class='op'>]</span><span class='op'>]</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'run_5102' not found</span></div><div class='input'>
<span class='va'>bam_path</span> <span class='op'>=</span> <span class='fu'><a href='createBamPath.html'>createBamPath</a></span><span class='op'>(</span><span class='fl'>5102</span>,
                         <span class='va'>wt_fastq</span>,
                         align_expr_prefix <span class='op'>=</span>
                         <span class='st'>"/mnt/htcf_scratch/chasem/rnaseq_pipeline/align_count_results"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in basename(fastq_filename): object 'wt_fastq' not found</span></div><div class='input'>
<span class='va'>cds_count</span> <span class='op'>=</span> <span class='fu'>countLibrary</span><span class='op'>(</span><span class='va'>bam_path</span>, <span class='va'>kn99_db</span>, <span class='st'>'cds'</span>, <span class='st'>'reverse'</span>, <span class='fl'>8</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'cdsBy': object 'kn99_db' not found</span></div><div class='input'>
<span class='va'>cds_counts_df</span> <span class='op'>=</span> <span class='fu'>as_tibble</span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>cds_counts</span><span class='op'>)</span>, rownames <span class='op'>=</span> <span class='st'>"gene_name"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in assay(cds_counts): could not find function "assay"</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


