<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ brentlabRnaSeqTools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">brentlabRnaSeqTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/create_nextflow_samplesheet.html">create_nextflow_samplesheet</a>
    </li>
    <li>
      <a href="../articles/create_set.html">create_set</a>
    </li>
    <li>
      <a href="../articles/database_interaction.html">database_interaction</a>
    </li>
    <li>
      <a href="../articles/experiment_sets.html">experiment_sets</a>
    </li>
    <li>
      <a href="../articles/qcExplorer.html">qcExplorer</a>
    </li>
    <li>
      <a href="../articles/running_the_nextflow_pipeline.html">running_the_nextflow_pipeline</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="addTablesToQcDatabase.html">addTablesToQcDatabase()</a></code> </p>
        </td>
        <td><p>add new_qc_database tables to old_qc_database_tables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="archiveDatabase.html">archiveDatabase()</a></code> </p>
        </td>
        <td><p>pull entire database (not counts) and save to output_dir for archival purposes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculateCoverage.html">calculateCoverage()</a></code> </p>
        </td>
        <td><p>calculate coverage over a given locus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculateGeneWiseMedians.html">calculateGeneWiseMedians()</a></code> </p>
        </td>
        <td><p>calculate medians across rows of dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculateRLE.html">calculateRLE()</a></code> </p>
        </td>
        <td><p>calculate RLE of a numeric dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compileQualimapSummaryMetrics.html">compileQualimapSummaryMetrics()</a></code> </p>
        </td>
        <td><p>parse qualimap summary output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compileTinOutput.html">compileTinOutput()</a></code> </p>
        </td>
        <td><p>extract all metrics from the output of tin.py run on a batch of files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="connectToDatabase.html">connectToDatabase()</a></code> </p>
        </td>
        <td><p>Connect to a remote postgresql database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countLibrary.html">countLibrary()</a></code> </p>
        </td>
        <td><p>Count reads in a library like HTSeq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countReadsInRanges.html">countReadsInRanges()</a></code> </p>
        </td>
        <td><p>Count reads over GRanges object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createBamPath.html">createBamPath()</a></code> </p>
        </td>
        <td><p>create a bam path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createDatabase.html">createDatabase()</a></code> </p>
        </td>
        <td><p>create the sqlite database as input to qcExplorer</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createEnvPertSet.html">createEnvPertSet()</a></code> </p>
        </td>
        <td><p>filter combined_df for environmental perturbation sample set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createInductionSetTally.html">createInductionSetTally()</a></code> </p>
        </td>
        <td><p>create 90 minute induction set tally</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createMergedTable.html">createMergedTable()</a></code> </p>
        </td>
        <td><p>Join QC tables (not qc_metrics) in qcExplorer input sqlite database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createNfCorePipelineSampleSheet.html">createNfCorePipelineSampleSheet()</a></code> </p>
        </td>
        <td><p>create nf-co sample sheet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createNinetyMinInductionModelMatrix.html">createNinetyMinInductionModelMatrix()</a></code> </p>
        </td>
        <td><p>Create the libraryProtocol + libraryDate model matrix with the earliest date of each library protocol dropped</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createNinetyMinuteInductionSet.html">createNinetyMinuteInductionSet()</a></code> </p>
        </td>
        <td><p>The current definition of the 90 minute induction dataset, according to the 2016 grant summary (loaded into environment, see head(grant_df)) -- single KO only</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createNinetyMinuteInductionWithDoubles.html">createNinetyMinuteInductionWithDoubles()</a></code> </p>
        </td>
        <td><p>The current definition of the 90 minute induction dataset, according to the 2016 grant summary (loaded into environment, see head(grant_df)) -- single and double KO</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createQCdatabase.html">createQCdatabase()</a></code> </p>
        </td>
        <td><p>create a sqlite database to hold the 'custom' qc data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createStatusDf.html">createStatusDf()</a></code> </p>
        </td>
        <td><p>create the metric status dataframe used to set thresholds and flag samples
in the QC explorer</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="database_colnames.html">database_colnames</a></code> </p>
        </td>
        <td><p>set global variables, mostly as a hack to make the tidyverse data binding pass the R CMD Check</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="database_info.html">database_info</a></code> </p>
        </td>
        <td><p>URLS to active databases</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="decomposeStatus2Bit.html">decomposeStatus2Bit()</a></code> </p>
        </td>
        <td><p>decompose sums of powers of two to a list of the summed powers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deseqObjectWithProtocolSpecificSizeFactors.html">deseqObjectWithProtocolSpecificSizeFactors()</a></code> </p>
        </td>
        <td><p>create deseq object with protocol specific size factors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="determineLibraryStrandedness.html">determineLibraryStrandedness()</a></code> </p>
        </td>
        <td><p>from the metadata libraryProtocol column, determine the library strandedness.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="edgerLog2cpm.html">edgerLog2cpm()</a></code> </p>
        </td>
        <td><p>Convenience function for edgeR's log2cpm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="examineSingleGroupWithLibDateSizeFactors.html">examineSingleGroupWithLibDateSizeFactors()</a></code> </p>
        </td>
        <td><p>temporary function to examine only PBS samples, eg</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="featureGRanges.html">featureGRanges()</a></code> </p>
        </td>
        <td><p>Given a GenomicFeatures annotation_db and a gene_id, extract an GRanges object of the cds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fltrLowReplicateParams.html">fltrLowReplicateParams()</a></code> </p>
        </td>
        <td><p>filter low replicate parameters from metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getBamIndexPath.html">getBamIndexPath()</a></code> </p>
        </td>
        <td><p>helper function to add .bai to bam path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCoverageOverRegion.html">getCoverageOverRegion()</a></code> </p>
        </td>
        <td><p>create a dataframe of coverage by nucleotide over a given locus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMetadata.html">getMetadata()</a></code> </p>
        </td>
        <td><p>Get the combined metadata as a tibble from a remote database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getNfCoMultiqcDataPaths.html">getNfCoMultiqcDataPaths()</a></code> </p>
        </td>
        <td><p>get relevant QC metrics tables from nf-co/rnaseq multiqc output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRawCounts.html">getRawCounts()</a></code> </p>
        </td>
        <td><p>Get combined raw counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRunNumberLeadingZero.html">getRunNumberLeadingZero()</a></code> </p>
        </td>
        <td><p>correct run number to add leading zero where approprirate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getUserAuthToken.html">getUserAuthToken()</a></code> </p>
        </td>
        <td><p>get (via a http POST request) your user authentication token from the database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="grant_df.html">grant_df</a></code> </p>
        </td>
        <td><p>the 2016 grant summary represented as a dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graphYeastTimeCourse.html">graphYeastTimeCourse()</a></code> </p>
        </td>
        <td><p>Plot time vs normalized counts of a given gene over n samples, faceted by librarydate and run</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="isNumeric.html">isNumeric()</a></code> </p>
        </td>
        <td><p>test if argument is numeric</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is_integer64.html">is_integer64()</a></code> </p>
        </td>
        <td><p>Test if value is datatype integer64</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listTables.html">listTables()</a></code> </p>
        </td>
        <td><p>list tables in databse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadAnnotationDatabase.html">loadAnnotationDatabase()</a></code> </p>
        </td>
        <td><p>Read in annotationbi tx_db</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="locusLog2Cpm.html">locusLog2Cpm()</a></code> </p>
        </td>
        <td><p>calculate log2cpm for a given locus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="outlierFence.html">outlierFence()</a></code> </p>
        </td>
        <td><p>calculate lower/upper fence as a default threshold on qc metrics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseBroadRnaseqcOutput.html">parseBroadRnaseqcOutput()</a></code> </p>
        </td>
        <td><p>extract all metrics from broad rnaseqc package output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseComparatives.html">parseComparatives()</a></code> </p>
        </td>
        <td><p>parse text comparative sentence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="patchTable.html">patchTable()</a></code> </p>
        </td>
        <td><p>PATCH entries in database table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCoverageOverLocus.html">plotCoverageOverLocus()</a></code> </p>
        </td>
        <td><p>plot coverage over locus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="postCounts.html">postCounts()</a></code> </p>
        </td>
        <td><p>post counts to database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="postFastqSheet.html">postFastqSheet()</a></code> </p>
        </td>
        <td><p>post new fastq sheet to database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="postQcSheet.html">postQcSheet()</a></code> </p>
        </td>
        <td><p>post new qc sheet to database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="postTable.html">postTable()</a></code> </p>
        </td>
        <td><p>Post a table to the database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="proteinCodingCount.html">proteinCodingCount()</a></code> </p>
        </td>
        <td><p>get total protein coding count from count dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcExplorer.html">qcExplorer()</a></code> </p>
        </td>
        <td><p>Explore QC metrics and set thresholds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcGenotypeAndMarkerCoverage.html">qcGenotypeAndMarkerCoverage()</a></code> </p>
        </td>
        <td><p>calcluate genotype and marker coverages for a given metadata df</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qualityAssessmentFilter.html">qualityAssessmentFilter()</a></code> </p>
        </td>
        <td><p>filter for manual passes (overrides auto fail) and automatic passes (unless auto failed)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="readInData.html">readInData()</a></code> </p>
        </td>
        <td><p>read in columnar data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeOneRedoIqr.html">removeOneRedoIqr()</a></code> </p>
        </td>
        <td><p>progressively remove max IQR sample and recalculate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeParameterEffects.html">removeParameterEffects()</a></code> </p>
        </td>
        <td><p>remove some effects from the counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rleByReplicateGroup.html">rleByReplicateGroup()</a></code> </p>
        </td>
        <td><p>calculate RLE by replicate groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rlePlot.html">rlePlot()</a></code> </p>
        </td>
        <td><p>plot RLE for a given column filter</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rlePlotCompareEffectRemoved.html">rlePlotCompareEffectRemoved()</a></code> </p>
        </td>
        <td><p>plots output of rleSummaryByReplicateGroup</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rlePlot_helper.html">rlePlot_helper()</a></code> </p>
        </td>
        <td><p>the actual plotting function for rlePlot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rleSummary.html">rleSummary()</a></code> </p>
        </td>
        <td><p>rleSummary calculates summary statistics of rleFullTable</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSVA.html">runSVA()</a></code> </p>
        </td>
        <td><p>run SVA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_numbers_with_leading_zero.html">run_numbers_with_leading_zero</a></code> </p>
        </td>
        <td><p>A named list containing a run number without a leading zero, eg 647, with the value being the same runnumber with
a leading 0, eg 0647.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="selectQaColumns.html">selectQaColumns()</a></code> </p>
        </td>
        <td><p>select fastqFileName, fastqFileNumber, and a pre-determined set of QC columns from a metadata df</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="strandedScanBamParam.html">strandedScanBamParam()</a></code> </p>
        </td>
        <td><p>create coverage scanbamparam object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="testBamPath.html">testBamPath()</a></code> </p>
        </td>
        <td><p>test bam path</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


