<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>create_set • brentlabRnaSeqTools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="create_set">
<meta property="og:description" content="brentlabRnaSeqTools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">brentlabRnaSeqTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/create_nextflow_samplesheet.html">create_nextflow_samplesheet</a>
    </li>
    <li>
      <a href="../articles/create_set.html">create_set</a>
    </li>
    <li>
      <a href="../articles/database_interaction.html">database_interaction</a>
    </li>
    <li>
      <a href="../articles/experiment_sets.html">experiment_sets</a>
    </li>
    <li>
      <a href="../articles/running_the_nextflow_pipeline.html">running_the_nextflow_pipeline</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="create_set_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>create_set</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/blob/master/vignettes/create_set.Rmd"><code>vignettes/create_set.Rmd</code></a></small>
      <div class="hidden name"><code>create_set.Rmd</code></div>

    </div>

    
    
<div id="most-of-what-you-need-will-be-loaded-in-brentlabrnaseqtools-already" class="section level2">
<h2 class="hasAnchor">
<a href="#most-of-what-you-need-will-be-loaded-in-brentlabrnaseqtools-already" class="anchor"></a>Most of what you need will be loaded in brentlabRnaSeqTools already</h2>
<p>But if you do get an error, read it carefully to see if it is simply that you need to load another package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cmatKhan/brentlabRnaSeqTools">brentlabRnaSeqTools</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="get-the-metadata-and-counts-from-the-database" class="section level2">
<h2 class="hasAnchor">
<a href="#get-the-metadata-and-counts-from-the-database" class="anchor"></a>get the metadata and counts from the database</h2>
<p>You will need to ask for the login credentials for this step. Storing them in your .Renviron makes it easy, but also make sure that you are careful not to push them up to a remote repo</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metadata_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/getMetadata.html">getMetadata</a></span><span class="op">(</span><span class="va">database_info</span><span class="op">$</span><span class="va">kn99_host</span>, 
                          <span class="va">database_info</span><span class="op">$</span><span class="va">kn99_db_name</span>, 
                          <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html">Sys.getenv</a></span><span class="op">(</span><span class="st">"DB_USERNAME"</span><span class="op">)</span>, 
                          <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html">Sys.getenv</a></span><span class="op">(</span><span class="st">"DB_PASSWORD"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># I do plan to make this part of the getMetaData() function, but it isn't yet</span>
<span class="va">metadata_df</span><span class="op">$</span><span class="va">fastqFileName</span> <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">metadata_df</span><span class="op">$</span><span class="va">fastqFileName</span>, <span class="st">".fastq.gz"</span><span class="op">)</span>

<span class="va">raw_counts_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/getRawCounts.html">getRawCounts</a></span><span class="op">(</span><span class="va">database_info</span><span class="op">$</span><span class="va">kn99_host</span>, 
                             <span class="va">database_info</span><span class="op">$</span><span class="va">kn99_db_name</span>, 
                             <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html">Sys.getenv</a></span><span class="op">(</span><span class="st">"DB_USERNAME"</span><span class="op">)</span>, 
                             <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html">Sys.getenv</a></span><span class="op">(</span><span class="st">"DB_PASSWORD"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># subset to only protein coding</span>
<span class="va">raw_counts_df</span> <span class="op">=</span> <span class="va">raw_counts_df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6967</span>,<span class="op">]</span></code></pre></div>
</div>
<div id="filter-the-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-the-metadata" class="anchor"></a>filter the metadata</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">subset_metadata</span> <span class="op">=</span> <span class="va">metadata_df</span> <span class="op">%&gt;%</span>
    <span class="co"># replace empty strings with NA</span>
    <span class="fu">mutate_if</span><span class="op">(</span><span class="va">is.character</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">~</span><span class="fu">na_if</span><span class="op">(</span><span class="va">.</span>,<span class="st">""</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># replace NAs with string entry</span>
    <span class="fu">mutate</span><span class="op">(</span>perturbation1 <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">perturbation1</span>, <span class="st">"noPerturbation"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>perturbation2 <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">perturbation2</span>, <span class="st">"noPerturbation"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>treatment <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">treatment</span>, <span class="st">'noTreatment'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>otherConditions <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">otherConditions</span>, <span class="st">'noOtherConditions'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>medium <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">medium</span>, <span class="st">'noMedium'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>atmosphere <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">atmosphere</span>, <span class="st">'noAtmosphere'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>pH <span class="op">=</span> <span class="fu">replace_na</span><span class="op">(</span><span class="va">pH</span>, <span class="st">'noPh'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>timePoint <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="va">timePoint</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">medium</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DMEM"</span><span class="op">)</span>,
           <span class="va">temperature</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">37</span><span class="op">)</span>,
           <span class="va">atmosphere</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CO2"</span><span class="op">)</span>,
           <span class="va">treatment</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"noTreatment"</span><span class="op">)</span>,
           <span class="va">otherConditions</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"noOtherConditions"</span><span class="op">)</span>,
           <span class="va">pH</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"noPh"</span><span class="op">)</span>,
           <span class="va">timePoint</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">90</span><span class="op">)</span>,
           <span class="va">strain</span> <span class="op">!=</span> <span class="st">"TDY1993"</span>,
           <span class="va">purpose</span><span class="op">==</span><span class="st">"fullRNASeq"</span>,
           <span class="va">perturbation1</span> <span class="op">==</span> <span class="st">"deletion"</span> <span class="op">|</span> <span class="va">perturbation1</span> <span class="op">==</span> <span class="st">"noPerturbation"</span>,
           <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">fastqFileName</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="look-at-what-we-got" class="section level2">
<h2 class="hasAnchor">
<a href="#look-at-what-we-got" class="anchor"></a>look at what we got</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">df</span> <span class="op">=</span> <span class="fu">tibble</span><span class="op">(</span>doubles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">genotype2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
            overexpression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">perturbation1</span> <span class="op">==</span> <span class="st">"over"</span><span class="op">)</span><span class="op">)</span>,
            singles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">genotype2</span><span class="op">)</span>, <span class="va">perturbation1</span> <span class="op">==</span> <span class="st">"deletion"</span><span class="op">)</span><span class="op">)</span>,
            wt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">genotype1</span> <span class="op">==</span> <span class="st">"CNAG_00000"</span><span class="op">)</span><span class="op">)</span>
            <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></code></pre></div>
<p>since I’m not evaluating this code for the vignette, this is what it looks like. If you copy and paste this code (after getting the login credentials, you can just run the code directly).</p>
<p>doubles: 116<br>
overexpression: 0 (purposefully filtered out due to linear dependence issue in model matrix) singles: 947<br>
wt: 238</p>
</div>
<div id="filter-for-autoaudit-status-though-override-this-if-manual-pass-true" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-for-autoaudit-status-though-override-this-if-manual-pass-true" class="anchor"></a>filter for autoAudit status, though override this if manual pass == TRUE</h2>
<p>NOTE! this function currently casts the column names to uppers. I am going to be removing this – previous, it was necessary b/c there were sometimes the same columns with different formatting. That can’t happen anymore</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">passing_subset_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/qualityAssessmentFilter.html">qualityAssessmentFilter</a></span><span class="op">(</span><span class="va">subset_metadata</span><span class="op">)</span>

<span class="va">passing_subset_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span><span class="op">(</span><span class="va">passing_subset_metadata</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">df</span> <span class="op">=</span> <span class="fu">tibble</span><span class="op">(</span>doubles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">passing_subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">GENOTYPE2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
            overexpression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">passing_subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">PERTURBATION1</span> <span class="op">==</span> <span class="st">"over"</span><span class="op">)</span><span class="op">)</span>,
            singles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">passing_subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">GENOTYPE2</span><span class="op">)</span>, <span class="va">PERTURBATION1</span> <span class="op">==</span> <span class="st">"deletion"</span><span class="op">)</span><span class="op">)</span>,
            wt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">nrow</a></span><span class="op">(</span><span class="va">passing_subset_metadata</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">GENOTYPE1</span> <span class="op">==</span> <span class="st">"CNAG_00000"</span><span class="op">)</span><span class="op">)</span>
            <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></code></pre></div>
<p>doubles: 73<br>
overexpression: 0<br>
singles: 658<br>
wt: 179</p>
</div>
<div id="figure-out-the-design" class="section level2">
<h2 class="hasAnchor">
<a href="#figure-out-the-design" class="anchor"></a>figure out the design</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">passing_subset_metadata_formatted</span> <span class="op">=</span> <span class="va">passing_subset_metadata</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>genotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">passing_subset_metadata</span><span class="op">$</span><span class="va">GENOTYPE1</span>, 
                          <span class="va">passing_subset_metadata</span><span class="op">$</span><span class="va">GENOTYPE2</span>,
                          sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>genotype <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">genotype</span>, <span class="st">"_NA"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>LIBRARYDATE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">LIBRARYDATE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>LIBRARYPROTOCOL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">LIBRARYPROTOCOL</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>genotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">genotype</span><span class="op">)</span><span class="op">)</span>

<span class="va">passing_subset_metadata_formatted</span><span class="op">$</span><span class="va">LIBRARYDATE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">passing_subset_metadata_formatted</span><span class="op">$</span><span class="va">LIBRARYDATE</span>,
                                              ref<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">passing_subset_metadata_formatted</span><span class="op">$</span><span class="va">LIBRARYDATE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">passing_subset_metadata_formatted</span><span class="op">$</span><span class="va">genotype</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">passing_subset_metadata_formatted</span><span class="op">$</span><span class="va">genotype</span>, 
                                                     ref<span class="op">=</span><span class="st">'CNAG_00000'</span><span class="op">)</span>

<span class="va">passing_subset_metadata_formatted</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span><span class="op">(</span><span class="va">passing_subset_metadata_formatted</span><span class="op">)</span></code></pre></div>
</div>
<div id="set-design-filter-out-samples-which-occur-in-parameters-with-low-replication" class="section level2">
<h2 class="hasAnchor">
<a href="#set-design-filter-out-samples-which-occur-in-parameters-with-low-replication" class="anchor"></a>set design, filter out samples which occur in parameters with low replication</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">design_formula</span> <span class="op">=</span> <span class="op">~</span><span class="va">LIBRARYPROTOCOL</span> <span class="op">+</span> <span class="va">LIBRARYDATE</span> <span class="op">+</span> <span class="va">genotype</span>

<span class="co"># filter out samples which occur in replicate sets of less than 2 in any of the parameters of the model matrix</span>
<span class="va">fltr_passing_subset_metadata_formatted</span> <span class="op">=</span> <span class="fu"><a href="../reference/fltrLowReplicateParams.html">fltrLowReplicateParams</a></span><span class="op">(</span><span class="va">passing_subset_metadata_formatted</span>, <span class="va">design_formula</span><span class="op">)</span></code></pre></div>
</div>
<div id="remove-high-dispersion-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-high-dispersion-genes" class="anchor"></a>remove high dispersion genes</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create list of high dispersion genes to filter out. Note: you'll need to get this from me for the time being</span>

<span class="va">old_protocol_high_disp_genes</span> <span class="op">=</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"~/projects/brentlab/90minuteInduction/datafreeze_20210528/data/old_protocol_high_disp_genes.csv"</span><span class="op">)</span><span class="op">$</span><span class="va">gene_ids</span>

<span class="va">new_protocol_high_disp_genes</span> <span class="op">=</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"~/projects/brentlab/90minuteInduction/datafreeze_20210528/data/new_protocol_high_disp_genes.csv"</span><span class="op">)</span><span class="op">$</span><span class="va">gene_ids</span>

<span class="va">high_disp_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">union</a></span><span class="op">(</span><span class="va">old_protocol_high_disp_genes</span>, <span class="va">new_protocol_high_disp_genes</span><span class="op">)</span>

<span class="va">full_gene_id_vector</span> <span class="op">=</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"~/Desktop/rnaseq_pipeline/rnaseq_pipeline/genome_files/KN99/KN99_gene_id_list.txt"</span>, col_names <span class="op">=</span> <span class="st">'gene_id'</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span>
<span class="va">full_gene_id_vector</span> <span class="op">=</span> <span class="va">full_gene_id_vector</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6967</span><span class="op">]</span>

<span class="va">fltr_passing_subset_counts</span> <span class="op">=</span> <span class="va">raw_counts_df</span><span class="op">[</span><span class="op">!</span><span class="va">full_gene_id_vector</span> <span class="op">%in%</span> <span class="va">high_disp_genes</span>,<span class="va">fltr_passing_subset_metadata_formatted</span><span class="op">$</span><span class="va">FASTQFILENAME</span><span class="op">]</span></code></pre></div>
</div>
<div id="attempt-to-make-dds-and-deal-with-linear-dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#attempt-to-make-dds-and-deal-with-linear-dependencies" class="anchor"></a>attempt to make dds and deal with linear dependencies</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dds</span> <span class="op">=</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">fltr_passing_subset_counts</span>,
                             colData <span class="op">=</span> <span class="va">fltr_passing_subset_metadata_formatted</span>,
                             design <span class="op">=</span> <span class="va">design_formula</span><span class="op">)</span></code></pre></div>
<p>Error in checkFullRank(modelMatrix) : the model matrix is not full rank, so the model cannot be fit as specified. One or more variables or interaction terms in the design formula are linear combinations of the others and must be removed. Please read the vignette section ‘Model matrix not full rank’: vignette(‘DESeq2’) 4. stop(“the model matrix is not full rank, so the model cannot be fit as specified.One or more variables or interaction terms in the design formula are linearcombinations of the others and must be removed.Please read the vignette section ‘Model matrix not full rank’:vignette(‘DESeq2’)”) 3. checkFullRank(modelMatrix) 2. DESeqDataSet(se, design = design, ignoreRank) 1. DESeqDataSetFromMatrix(countData = fltr_passing_subset_counts, colData = fltr_passing_subset_metadata_formatted, design = design_formula)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">caret</span><span class="op">)</span>
<span class="va">model_matrix</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="va">design_formula</span>, <span class="va">fltr_passing_subset_metadata_formatted</span><span class="op">)</span>

<span class="va">caret_output</span> <span class="op">=</span> <span class="fu">findLinearCombos</span><span class="op">(</span><span class="va">model_matrix</span><span class="op">)</span></code></pre></div>
<p>$linearCombos $linearCombos[[1]] [1] 59 1 2 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58</p>
<p>$remove [1] 59</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># we are filtering out this: "LIBRARYDATE2021-05-06", which is a 'new protocol' library date. Note that a</span>
<span class="co"># 'old protocol' date was also dropped</span>
<span class="va">augment_model_matrix</span> <span class="op">=</span> <span class="va">model_matrix</span><span class="op">[</span>, <span class="op">-</span><span class="fl">59</span><span class="op">]</span>

<span class="fu">findLinearCombos</span><span class="op">(</span><span class="va">augment_model_matrix</span><span class="op">)</span></code></pre></div>
<p>$linearCombos list()</p>
<p>$remove NULL ## get protocol specific size factors</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># note: i am going to re-do this function so it only gives back the size factors. for now, it is purpose built for the 90minuteInduction</span>
<span class="va">size_factor_dds</span> <span class="op">=</span> <span class="fu"><a href="../reference/deseqObjectWithProtocolSpecificSizeFactors.html">deseqObjectWithProtocolSpecificSizeFactors</a></span><span class="op">(</span><span class="va">fltr_passing_subset_metadata_formatted</span>, <span class="va">fltr_passing_subset_counts</span><span class="op">)</span>

<span class="va">size_factor_df</span> <span class="op">=</span> <span class="fu">tibble</span><span class="op">(</span>FASTQFILENAME <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>, size_factors <span class="op">=</span> <span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>

<span class="va">fltr_passing_subset_metadata_formatted_with_sizeFactors</span> <span class="op">=</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">fltr_passing_subset_metadata_formatted</span>, <span class="va">size_factor_df</span>, by<span class="op">=</span><span class="st">'FASTQFILENAME'</span><span class="op">)</span></code></pre></div>
</div>
<div id="check-that-the-rows-and-columns-are-in-the-right-order" class="section level2">
<h2 class="hasAnchor">
<a href="#check-that-the-rows-and-columns-are-in-the-right-order" class="anchor"></a>check that the rows and columns are in the right order</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">colnames</a></span><span class="op">(</span><span class="va">fltr_passing_subset_counts</span><span class="op">)</span>, <span class="va">fltr_passing_subset_metadata_formatted_with_sizeFactors</span><span class="op">$</span><span class="va">FASTQFILENAME</span><span class="op">)</span></code></pre></div>
</div>
<div id="write-out-the-dds-for-first-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#write-out-the-dds-for-first-processing" class="anchor"></a>write out the dds for first processing</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dds</span> <span class="op">=</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">fltr_passing_subset_counts</span>,
                               colData <span class="op">=</span> <span class="va">fltr_passing_subset_metadata_formatted_with_sizeFactors</span>,
                               design <span class="op">=</span> <span class="va">augment_model_matrix</span><span class="op">)</span>

<span class="va">output_dir</span> <span class="op">=</span> <span class="st">"/mnt/htcf_scratch/chasem/rnaseq_pipeline/experiments/20210610/frankenstein_set"</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">output_dir</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">write_rds</span><span class="op">(</span><span class="va">dds</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>,<span class="st">"input_dds_before_rle.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="read-in-the-processed-dds" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-the-processed-dds" class="anchor"></a>read in the processed dds</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">output_dir</span> <span class="op">=</span> <span class="st">"/mnt/htcf_scratch/chasem/rnaseq_pipeline/experiments/20210610/frankenstein_set"</span>
<span class="va">filename</span> <span class="op">=</span> <span class="st">"deseq_output.rds"</span>

<span class="va">dds</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="va">filename</span><span class="op">)</span><span class="op">)</span>

<span class="va">metadata_df</span> <span class="op">=</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<div id="extract-norm-counts" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-norm-counts" class="anchor"></a>extract norm counts</h3>
</div>
</div>
<div id="remove-the-library-protocol-and-date-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-the-library-protocol-and-date-effects" class="anchor"></a>remove the library protocol and date effects</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">start</span><span class="op">=</span><span class="fl">2</span>
<span class="va">stop</span><span class="op">=</span><span class="fl">58</span>
<span class="va">batch_indicies</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="va">start</span>,<span class="va">stop</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="fu">design</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">batch_indicies</span><span class="op">]</span></code></pre></div>
</div>
<div id="calculate-rle-by-replicate-group" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-rle-by-replicate-group" class="anchor"></a>calculate rle by replicate group</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">replicate_split</span> <span class="op">=</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">group_by</span><span class="op">(</span><span class="va">genotype</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">group_split</span><span class="op">(</span><span class="op">)</span>

<span class="va">replicate_names</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">replicate_split</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu">pull</span><span class="op">(</span><span class="va">x</span>,<span class="va">genotype</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">replicate_sample_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">replicate_split</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu">pull</span><span class="op">(</span><span class="va">x</span>,<span class="va">FASTQFILENAME</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">replicate_sample_list</span><span class="op">)</span> <span class="op">=</span> <span class="va">replicate_names</span>

<span class="va">norm_counts_rle_by_replicate_group</span> <span class="op">=</span> <span class="fu"><a href="../reference/rleByReplicateGroup.html">rleByReplicateGroup</a></span><span class="op">(</span><span class="va">replicate_sample_list</span>, 
                                                                <span class="va">norm_counts</span>, 
                                                                log2_transformed_flag <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
  
<span class="va">norm_counts_rle_summary</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">norm_counts_rle_by_replicate_group</span>, <span class="va">rleSummary</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">left_join</span><span class="op">(</span><span class="va">metadata_df</span>, by<span class="op">=</span><span class="st">'FASTQFILENAME'</span><span class="op">)</span>

<span class="va">removed_effect_rle_by_replicate_group</span> <span class="op">=</span> <span class="fu"><a href="../reference/rleByReplicateGroup.html">rleByReplicateGroup</a></span><span class="op">(</span><span class="va">replicate_sample_list</span>, 
                                                                   <span class="va">removed_librarydate_effect_log_norm_counts</span>, 
                                                                   log2_transformed_flag <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">removed_effect_rle_summary</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">removed_effect_rle_by_replicate_group</span>, <span class="va">rleSummary</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">left_join</span><span class="op">(</span><span class="va">metadata_df</span>, by<span class="op">=</span><span class="st">'FASTQFILENAME'</span><span class="op">)</span></code></pre></div>
</div>
<div id="look-at-the-iqr-dist" class="section level2">
<h2 class="hasAnchor">
<a href="#look-at-the-iqr-dist" class="anchor"></a>look at the iqr dist</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plotRLE_bar_iqr</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">rle_summary</span>, <span class="va">title</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="va">rle_summary</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">INTERQUARTILE_RANGE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="va">title</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">hist</span> <span class="op">=</span> <span class="fu">plotRLE_bar_iqr</span><span class="op">(</span><span class="va">removed_effect_rle_summary</span>, <span class="st">"iqr"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">hist</span><span class="op">)</span></code></pre></div>
</div>
<div id="get-high-rle-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#get-high-rle-plots" class="anchor"></a>get high RLE plots</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">threshold</span> <span class="op">=</span> <span class="fl">.6125</span>

<span class="va">high_iqr_genotype_groups</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">removed_effect_rle_summary</span> <span class="op">%&gt;%</span> 
                                  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">INTERQUARTILE_RANGE</span> <span class="op">&gt;</span> <span class="va">threshold</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                                  <span class="fu">left_join</span><span class="op">(</span><span class="va">removed_effect_rle_summary</span>, by<span class="op">=</span><span class="st">'GENOTYPE1'</span><span class="op">)</span> <span class="op">%&gt;%</span>
                                  <span class="fu">group_by</span><span class="op">(</span><span class="va">GENOTYPE1</span><span class="op">)</span> <span class="op">%&gt;%</span>
                                  <span class="fu">pull</span><span class="op">(</span><span class="va">GENOTYPE1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="look-at-high-rle-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#look-at-high-rle-plots" class="anchor"></a>look at high rle plots</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">high_iqr_genotype_groups</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/rlePlotCompareEffectRemoved.html">rlePlotCompareEffectRemoved</a></span><span class="op">(</span><span class="va">norm_counts_rle_by_replicate_group</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,
                                                                         <span class="va">removed_effect_rle_by_replicate_group</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, 
                                                                         <span class="va">metadata_df</span>,
                                                                         <span class="va">x</span><span class="op">)</span><span class="op">)</span>

<span class="va">subset_samples</span> <span class="op">=</span> <span class="va">metadata_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">LIBRARYDATE</span> <span class="op">==</span> <span class="st">'2013-04-17'</span>, <span class="va">GENOTYPE</span> <span class="op">==</span> <span class="st">"CNAG_00000"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">FASTQFILENAME</span><span class="op">)</span>

<span class="fu"><a href="../reference/rlePlotCompareEffectRemoved.html">rlePlotCompareEffectRemoved</a></span><span class="op">(</span><span class="va">norm_counts_rle_by_replicate_group</span><span class="op">$</span><span class="va">CNAG_00000</span><span class="op">[</span>,<span class="va">subset_samples</span><span class="op">]</span>, 
                            <span class="va">removed_effect_rle_by_replicate_group</span><span class="op">$</span><span class="va">CNAG_00000</span><span class="op">[</span>,<span class="va">subset_samples</span><span class="op">]</span>,
                            <span class="va">metadata_df</span>,
                            <span class="st">'cnag_00000'</span><span class="op">)</span></code></pre></div>
</div>
<div id="filter-by-rle" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-by-rle" class="anchor"></a>filter by rle</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">iqr_fltr_rle_summary</span> <span class="op">=</span> <span class="va">removed_effect_rle_summary</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">INTERQUARTILE_RANGE</span><span class="op">&lt;</span><span class="va">threshold</span><span class="op">)</span>

<span class="fu">plotRLE_bar_iqr</span><span class="op">(</span><span class="va">iqr_fltr_rle_summary</span>, <span class="st">"iqr"</span><span class="op">)</span></code></pre></div>
</div>
<div id="remove-old-size-factors" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-old-size-factors" class="anchor"></a>remove old size factors</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">iqr_fltr_rle_summary</span> <span class="op">=</span> <span class="va">iqr_fltr_rle_summary</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">sizeFactor</span><span class="op">)</span></code></pre></div>
</div>
<div id="remove-samples-that-fall-in-parameters-with-low-replicate-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-samples-that-fall-in-parameters-with-low-replicate-counts" class="anchor"></a>remove samples that fall in parameters with low replicate counts</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">iqr_fltr_meta</span> <span class="op">=</span> <span class="fu"><a href="../reference/fltrLowReplicateParams.html">fltrLowReplicateParams</a></span><span class="op">(</span><span class="va">iqr_fltr_rle_summary</span>, <span class="fu">design</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="fu">design</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, <span class="va">iqr_fltr_meta</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>note: if samples are removed, it is likely that you’ll need to check for linear dependencies in the model matrix again</p>
</div>
<div id="filter-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-counts" class="anchor"></a>filter counts</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">iqr_fltr_counts</span> <span class="op">=</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">[</span>,<span class="va">iqr_fltr_meta</span><span class="op">$</span><span class="va">FASTQFILENAME</span><span class="op">]</span></code></pre></div>
</div>
<div id="check-that-the-rows-and-columns-are-right" class="section level2">
<h2 class="hasAnchor">
<a href="#check-that-the-rows-and-columns-are-right" class="anchor"></a>check that the rows and columns are right</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="va">iqr_fltr_meta</span><span class="op">$</span><span class="va">FASTQFILENAME</span>, <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html">colnames</a></span><span class="op">(</span><span class="va">iqr_fltr_counts</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="recalculate-size-factors" class="section level2">
<h2 class="hasAnchor">
<a href="#recalculate-size-factors" class="anchor"></a>recalculate size factors</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># note: i am going to re-do this function so it only gives back the size factors. for now, it is purpose built for the 90minuteInduction</span>
<span class="va">size_factor_dds</span> <span class="op">=</span> <span class="fu"><a href="../reference/deseqObjectWithProtocolSpecificSizeFactors.html">deseqObjectWithProtocolSpecificSizeFactors</a></span><span class="op">(</span><span class="va">iqr_fltr_rle_summary</span>, <span class="va">fltr_passing_subset_counts</span><span class="op">)</span>

<span class="va">size_factor_df</span> <span class="op">=</span> <span class="fu">tibble</span><span class="op">(</span>FASTQFILENAME <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>, size_factors <span class="op">=</span> <span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>

<span class="va">fltr_passing_subset_metadata_formatted_with_sizeFactors</span> <span class="op">=</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">fltr_passing_subset_metadata_formatted</span>, <span class="va">size_factor_df</span>, by<span class="op">=</span><span class="st">'FASTQFILENAME'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>revised_dds <span class="ot">=</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> iqr_fltr_counts,</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">colData =</span> iqr_fltr_meta,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">design =</span> <span class="co"># note -- this may need to be re-done after iqr filtering)</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># write_rds(revised_dds, "/mnt/htcf_scratch/chasem/rnaseq_pipeline/experiments/20210605/other_regulators/after_rle/other_regulators_after_rle_deseq_input.rds")</span></span></code></pre></div>
</div>
<div id="sva" class="section level2">
<h2 class="hasAnchor">
<a href="#sva" class="anchor"></a>SVA</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sva)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">MulticoreParam</span>(<span class="dv">3</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>sva_func <span class="ot">=</span> <span class="cf">function</span>(num_sv){</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">svaseq</span>(<span class="fu">as.matrix</span>(iqr_fltr_counts),</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">mod=</span><span class="co"># note -- this may need to be re-done after iqr filtering,</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">mod0=</span><span class="co"># note -- this may need to be re-done after iqr filtering,</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">n.sv=</span>num_sv)</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a><span class="co"># note: the first 1-58 is intercept to the last batch index</span></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a><span class="co">#sva_output_list = bplapply(1:3, sva_func) --&gt; when trying, kept getting error about computationally singular matrix</span></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="co"># when trying without specifying n.sv, the estimate was 22 SV</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>sv_1 <span class="ot">=</span> <span class="fu">sva_func</span>(<span class="dv">1</span>)</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>sv_2 <span class="ot">=</span> <span class="fu">sva_func</span>(<span class="dv">2</span>)</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>sv_3 <span class="ot">=</span> <span class="fu">sva_func</span>(<span class="dv">3</span>)</span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dds_list = list(</span>
<span class="co">#   dds_0 = DESeqDataSetFromMatrix(countData = fltr_passing_subset_counts,</span>
<span class="co">#                                colData = fltr_passing_subset_metadata_formatted,</span>
<span class="co">#                                design = augment_model_matrix),</span>
<span class="co">#   dds_1 = DESeqDataSetFromMatrix(countData = fltr_passing_subset_counts,</span>
<span class="co">#                                colData = fltr_passing_subset_metadata_formatted,</span>
<span class="co">#                                design = augment_model_matrix_sv1),</span>
<span class="co">#   dds_2 = DESeqDataSetFromMatrix(countData = fltr_passing_subset_counts,</span>
<span class="co">#                                colData = fltr_passing_subset_metadata_formatted,</span>
<span class="co">#                                design = augment_model_matrix_sv2)</span>
<span class="co">#   dds_3</span>
<span class="co"># </span>
<span class="co"># )</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
