<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>genomics_tools • brentlabRnaSeqTools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="genomics_tools">
<meta property="og:description" content="brentlabRnaSeqTools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">brentlabRnaSeqTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/create_nextflow_samplesheet.html">create_nextflow_samplesheet</a>
    </li>
    <li>
      <a href="../articles/create_set.html">create_set</a>
    </li>
    <li>
      <a href="../articles/database_interaction.html">database_interaction</a>
    </li>
    <li>
      <a href="../articles/experiment_sets.html">experiment_sets</a>
    </li>
    <li>
      <a href="../articles/genomics_tools.html">genomics_tools</a>
    </li>
    <li>
      <a href="../articles/install_nf_co_rnaseq_pipeline.html">install_nf_co_rnaseq_pipeline</a>
    </li>
    <li>
      <a href="../articles/running_the_nextflow_pipeline.html">running_the_nextflow_pipeline</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="genomics_tools_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>genomics_tools</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/blob/master/vignettes/genomics_tools.Rmd"><code>vignettes/genomics_tools.Rmd</code></a></small>
      <div class="hidden name"><code>genomics_tools.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cmatKhan/brentlabRnaSeqTools">brentlabRnaSeqTools</a></span><span class="op">)</span></code></pre></div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p>This is meant as an index of genomics tools that I have found and tried. I am starting this on 20210809 and will try to keep it up to date from now on</p>
</div>
<div id="alignments" class="section level1">
<h1 class="hasAnchor">
<a href="#alignments" class="anchor"></a>Alignments</h1>
<p>readGAlignments and GenomicFeatures work. However, I find GenomicFeatures to be both difficult to use and buggy unless you are using a genome from AnnotateDbi or the like. My personal preference right now is using <code>rtracklayer::import</code> to create a gRangeList object from an annotation file. You can use that to visualize using the function in this package called <code>createIgvBatchScript</code>, or anything else that takes granges as input.</p>
</div>
<div id="counts" class="section level1">
<h1 class="hasAnchor">
<a href="#counts" class="anchor"></a>Counts</h1>
<p>Rsubreads is <strong>very</strong> fast. It is a raw count algorithm, similar to HTSeq (but faster).</p>
</div>
<div id="browser-and-browser-esque-visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#browser-and-browser-esque-visualization" class="anchor"></a>Browser and browser-esque visualization</h1>
<div id="igv" class="section level2">
<h2 class="hasAnchor">
<a href="#igv" class="anchor"></a>IGV</h2>
<p>As far as I know, there is no good interactive R IGV package. There is a function in this package, <code>createIgvBatchScript</code>, which will create a batch script for you. You can use this with <code>map</code> or <code>lapply</code> to create any number of batch scripts, and then run those separate scripts either on a cluster as an array job or sequentially on your local. See <code>?createIgvBatchScript</code>.</p>
</div>
<div id="ggbio" class="section level2">
<h2 class="hasAnchor">
<a href="#ggbio" class="anchor"></a>ggbio</h2>
<p>This works great – it uses the bioconductor genomics objects (GRanges, annotation databases), and produces good enough visualizations (igv/ucsc is better, but this is great/better for quickly producing plots)</p>
</div>
<div id="wiggleplotr" class="section level2">
<h2 class="hasAnchor">
<a href="#wiggleplotr" class="anchor"></a>wiggleplotr</h2>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/wiggleplotr/inst/doc/wiggleplotr.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/wiggleplotr/inst/doc/wiggleplotr.html</a></p>
<p>For transcript visualization. this works quite nicely. needs bigWig files (produced by the nf-co/rnaseq pipeline). Can use txDb to create GRangesList of transcripts to a given gene</p>
<p>This is a dummy example – this plots separate gene CDS rather than transcripts of a single gene – but it works for crypto with the txDb created from the gtf</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">some_cds</span> <span class="op">=</span> <span class="va">kn99_cds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CNAG_02196"</span>, <span class="st">"CNAG_02197"</span>,<span class="st">"CNAG_02198"</span><span class="op">)</span><span class="op">]</span>
<span class="fu">plotTranscripts</span><span class="op">(</span><span class="va">some_cds</span><span class="op">)</span></code></pre></div>
</div>
<div id="rtracklayer" class="section level2">
<h2 class="hasAnchor">
<a href="#rtracklayer" class="anchor"></a>rtracklayer</h2>
<p>If you’re using a model organism that already has the annotation/genome object created by UCSC (eg human), then this works. Note that as of 2021 11 11, you need an absolutely up to date version of BiocManager for <code>hg38</code> to work on the UCSC browser.</p>
</div>
<div id="trackplot" class="section level2">
<h2 class="hasAnchor">
<a href="#trackplot" class="anchor"></a>trackplot</h2>
<p><a href="https://github.com/PoisonAlien/trackplot">https://github.com/PoisonAlien/trackplot</a></p>
<p>This did not work for crypto due to some malformed GTF entries – see error:</p>
<pre class="raw"><code>
&gt; track_plot(summary_list = track_data, draw_gene_track = TRUE, gene_model = "data/liftoff_h99_to_kn99_with_markers.gtf", isGTF = TRUE)
Parsing gtf file..
Error in data.table::foverlaps(x = query, y = gtf, type = "any", nomatch = NULL) : 
  NA values in data.table 'y' start column: 'start'. All rows with NA values in the range columns must be removed for foverlaps() to work.
</code></pre>
<p>However, it probably would work for human data. The visualizations look nice, and the nf-co/rnaseq pipeline produces bigwig files.</p>
<p>The one dependency is bwtools. Installing bwtools was a challenge – the author of trackplot had a useful comment in the issues, however:</p>
<p><a href="https://github.com/CRG-Barcelona/bwtool/issues/49#issuecomment-604477559">This worked for me</a>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone <span class="st">'https://github.com/CRG-Barcelona/bwtool'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone <span class="st">'https://github.com/CRG-Barcelona/libbeato'</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/madler/zlib</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> libbeato/</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> checkout 0c30432af9c7e1e09ba065ad3b2bc042baa54dc2</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="ex">./configure</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> make install <span class="co"># do this to install globally. omit if installing locally, eg on the cluster</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> zlib</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="ex">./configure</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> make install <span class="co"># see above</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> bwtool/</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co"># if you have installed this globally with sudo</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="ex">./configure</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="co"># if you are installing localling, then do this:</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="ex">./configure</span> CFLAGS=<span class="st">'-I../libbeato -I../zlib'</span> LDFLAGS=<span class="st">'-L../libbeato/jkweb -L../libbeato/beato -L../zlib'</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="co"># if installing globally</span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> make install</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
