<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>experiment_sets • brentlabRnaSeqTools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="experiment_sets">
<meta property="og:description" content="brentlabRnaSeqTools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">brentlabRnaSeqTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/create_nextflow_samplesheet.html">create_nextflow_samplesheet</a>
    </li>
    <li>
      <a href="../articles/create_set.html">create_set</a>
    </li>
    <li>
      <a href="../articles/database_interaction.html">database_interaction</a>
    </li>
    <li>
      <a href="../articles/experiment_sets.html">experiment_sets</a>
    </li>
    <li>
      <a href="../articles/genomics_tools.html">genomics_tools</a>
    </li>
    <li>
      <a href="../articles/install_nf_co_rnaseq_pipeline.html">install_nf_co_rnaseq_pipeline</a>
    </li>
    <li>
      <a href="../articles/running_the_nextflow_pipeline.html">running_the_nextflow_pipeline</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="experiment_sets_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>experiment_sets</h1>
            
            <h4 class="date">11/12/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/blob/master/vignettes/experiment_sets.Rmd"><code>vignettes/experiment_sets.Rmd</code></a></small>
      <div class="hidden name"><code>experiment_sets.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Explanations of provided experiment set filters, and some examples</p>
    </div>
    
<div id="grant-summary" class="section level1">
<h1 class="hasAnchor">
<a href="#grant-summary" class="anchor"></a>2016 grant summary</h1>
<p>The 2016 grant summary is stored in a variable, <code>grant_df</code> which is available when you execute <code><a href="https://github.com/cmatKhan/brentlabRnaSeqTools">library(brentlabRnaSeqTools)</a></code></p>
</div>
<div id="general-notes" class="section level1">
<h1 class="hasAnchor">
<a href="#general-notes" class="anchor"></a>General notes</h1>
<p>In general, the first step to creating an experimental set is to ensure that all ‘empty’ values in the data are the same. This may be accomplished by setting any empty strings to NA like so (using dplyr):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">metadata</span> <span class="op">=</span> <span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="co"># replace empty strings with NA</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_if</a></span><span class="op">(</span><span class="va">is.character</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/na_if.html">na_if</a></span><span class="op">(</span><span class="va">.</span>,<span class="st">""</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Once this is accomplished, you may easily replace the NAs with a specific entry. An example of this will be in the text of the <code><a href="../reference/createNinetyMinuteInductionSet.html">createNinetyMinuteInductionSet()</a></code> function.</p>
<p>Filtering with dplyr is quite simple. Extensive documentation may be found here:</p>
<p><a href="https://dplyr.tidyverse.org/" class="uri">https://dplyr.tidyverse.org/</a></p>
<p>And you may also use the functions listed below as templates.</p>
<p>One handy trick: use the %in% operator to filter on multiple values in a given field. For example:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">temperature</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">37</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>would return any record with temperature either 37 or 30.</p>
<p>Another trick is to use grepl to search for a pattern in a field. For example:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"PLAG"</span>, <span class="va">s2cDNAPreparer</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>which may simiarly be achived with</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html">str_detect</a></span><span class="op">(</span><span class="va">s2cDNAPreparer</span>, <span class="st">"PLAG"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="minuteinduction---2016-grant-only" class="section level1">
<h1 class="hasAnchor">
<a href="#minuteinduction---2016-grant-only" class="anchor"></a>90minuteInduction - 2016 grant only</h1>
<p>The definition of the 90 minute induction set is:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">createNinetyMinuteInductionSet</span>
<span class="co">#&gt; function (metadata, grant_df) </span>
<span class="co">#&gt; {</span>
<span class="co">#&gt;     metadata = metadata %&gt;% mutate_if(is.character, list(~na_if(., </span>
<span class="co">#&gt;         ""))) %&gt;% mutate(treatment = replace_na(treatment, "noTreatment")) %&gt;% </span>
<span class="co">#&gt;         mutate(otherConditions = replace_na(otherConditions, </span>
<span class="co">#&gt;             "noOtherConditions")) %&gt;% mutate(medium = replace_na(medium, </span>
<span class="co">#&gt;         "noMedium")) %&gt;% mutate(atmosphere = replace_na(atmosphere, </span>
<span class="co">#&gt;         "noAtmosphere")) %&gt;% mutate(pH = replace_na(pH, "noPh"))</span>
<span class="co">#&gt;     metadata$temperature %&lt;&gt;% as.numeric</span>
<span class="co">#&gt;     metadata$timePoint %&lt;&gt;% as.numeric</span>
<span class="co">#&gt;     condition_fltr_metadata = metadata %&gt;% filter(medium %in% </span>
<span class="co">#&gt;         c("DMEM"), temperature %in% c(37), atmosphere %in% c("CO2"), </span>
<span class="co">#&gt;         treatment %in% c("noTreatment"), otherConditions %in% </span>
<span class="co">#&gt;             c("noOtherConditions"), pH %in% c("noPh"), timePoint %in% </span>
<span class="co">#&gt;             c(90), strain != "TDY1993", purpose == "fullRNASeq", </span>
<span class="co">#&gt;         !is.na(fastqFileName), str_detect(genotype1, "CNAG"), </span>
<span class="co">#&gt;         is.na(genotype2), perturbation1 == "deletion" | is.na(perturbation1))</span>
<span class="co">#&gt;     wt_induction_set = condition_fltr_metadata %&gt;% filter(genotype1 == </span>
<span class="co">#&gt;         "CNAG_00000", strain == "TDY451")</span>
<span class="co">#&gt;     fltr_grant_df = grant_df %&gt;% filter(STRAIN_STATUS == "done")</span>
<span class="co">#&gt;     perturbed_induction_set = condition_fltr_metadata %&gt;% filter(genotype1 %in% </span>
<span class="co">#&gt;         fltr_grant_df$GENOTYPE1 &amp; is.na(genotype2))</span>
<span class="co">#&gt;     induction_set = bind_rows(wt_induction_set, perturbed_induction_set)</span>
<span class="co">#&gt;     colnames(induction_set) = toupper(colnames(induction_set))</span>
<span class="co">#&gt;     induction_set$FASTQFILENAME = str_remove(induction_set$FASTQFILENAME, </span>
<span class="co">#&gt;         ".fastq.gz")</span>
<span class="co">#&gt;     return(induction_set)</span>
<span class="co">#&gt; }</span>
<span class="co">#&gt; &lt;bytecode: 0x55a454517a68&gt;</span>
<span class="co">#&gt; &lt;environment: namespace:brentlabRnaSeqTools&gt;</span></code></pre></div>
</div>
<div id="minuteinduction---with-doubles" class="section level1">
<h1 class="hasAnchor">
<a href="#minuteinduction---with-doubles" class="anchor"></a>90minuteInduction - with doubles</h1>
<p>This includes all double deletions of genotypes which are included in the 2016 grant summary</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">createNinetyMinuteInductionWithDoubles</span>
<span class="co">#&gt; function (metadata, grant_df) </span>
<span class="co">#&gt; {</span>
<span class="co">#&gt;     metadata = metadata %&gt;% mutate(treatment = replace_na(treatment, </span>
<span class="co">#&gt;         "noTreatment")) %&gt;% mutate(otherConditions = replace_na(otherConditions, </span>
<span class="co">#&gt;         "noOtherConditions")) %&gt;% mutate(medium = replace_na(medium, </span>
<span class="co">#&gt;         "noMedium")) %&gt;% mutate(atmosphere = replace_na(atmosphere, </span>
<span class="co">#&gt;         "noAtmosphere")) %&gt;% mutate(pH = replace_na(pH, "noPh"))</span>
<span class="co">#&gt;     metadata$temperature %&lt;&gt;% as.numeric</span>
<span class="co">#&gt;     metadata$timePoint %&lt;&gt;% as.numeric</span>
<span class="co">#&gt;     condition_fltr_metadata = metadata %&gt;% filter(medium %in% </span>
<span class="co">#&gt;         c("DMEM"), temperature %in% c(37), atmosphere %in% c("CO2"), </span>
<span class="co">#&gt;         treatment %in% c("noTreatment"), otherConditions %in% </span>
<span class="co">#&gt;             c("noOtherConditions"), pH %in% c("noPh"), timePoint %in% </span>
<span class="co">#&gt;             c(90), strain != "TDY1993", purpose == "fullRNASeq", </span>
<span class="co">#&gt;         !is.na(fastqFileName), str_detect(genotype1, "CNAG"), </span>
<span class="co">#&gt;         perturbation1 == "deletion" | is.na(perturbation1), perturbation2 == </span>
<span class="co">#&gt;             "deletion" | is.na(perturbation2))</span>
<span class="co">#&gt;     wt_induction_set = condition_fltr_metadata %&gt;% filter(genotype1 == </span>
<span class="co">#&gt;         "CNAG_00000", strain == "TDY451")</span>
<span class="co">#&gt;     perturbed_induction_set = condition_fltr_metadata %&gt;% filter(genotype1 %in% </span>
<span class="co">#&gt;         grant_df$GENOTYPE1 &amp; (genotype2 %in% grant_df$GENOTYPE1 | </span>
<span class="co">#&gt;         is.na(genotype2)))</span>
<span class="co">#&gt;     induction_set = bind_rows(wt_induction_set, perturbed_induction_set)</span>
<span class="co">#&gt;     colnames(induction_set) = toupper(colnames(induction_set))</span>
<span class="co">#&gt;     induction_set$FASTQFILENAME = str_remove(induction_set$FASTQFILENAME, </span>
<span class="co">#&gt;         ".fastq.gz")</span>
<span class="co">#&gt;     return(induction_set)</span>
<span class="co">#&gt; }</span>
<span class="co">#&gt; &lt;bytecode: 0x55a48b01f618&gt;</span>
<span class="co">#&gt; &lt;environment: namespace:brentlabRnaSeqTools&gt;</span></code></pre></div>
<p>To create a set with all doubles in the database, simply filter for samples which have a double deletion. For example:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># see database_interaction for instructions on creating the combined_df</span>
<span class="va">combined_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">perturbation1</span><span class="op">==</span><span class="st">'deletion'</span> <span class="op">&amp;</span> <span class="va">perturbation2</span><span class="op">==</span><span class="st">'deletion'</span><span class="op">)</span>

<span class="co"># at this point, you may choose to filter further. For example, you may use the same conditions as the 90minuteInduction</span>
<span class="co"># see createNinetyMinuteInductionSet for those conditions</span></code></pre></div>
</div>
<div id="minuteinduction---other-regulators" class="section level1">
<h1 class="hasAnchor">
<a href="#minuteinduction---other-regulators" class="anchor"></a>90minuteInduction - other regulators</h1>
<p>This is the set of single perturbations which are not included in the 2016 grant summary</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get wildtypes from the createNinetyMinuteInduction set</span>
<span class="va">grant_induction_set</span> <span class="op">=</span> <span class="fu"><a href="../reference/createNinetyMinuteInductionSet.html">createNinetyMinuteInductionSet</a></span><span class="op">(</span><span class="va">combined_df</span>, <span class="va">grant_df</span><span class="op">)</span>

<span class="va">induction_set_wt</span> <span class="op">=</span> <span class="va">grant_induction_set</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">genotype1</span> <span class="op">==</span> <span class="st">'CNAG_00000'</span><span class="op">)</span>

<span class="co"># note: this isn't as explicit as the createNinetyMinuteInductionSet definition. It likely should be, which would mean replacing empty strings with a definite value, and then filtering exlicitly in those columns. This prevents any strange behavior with empty strings and NA</span>
<span class="va">regulators_not_in_2016_grant</span> <span class="op">=</span> <span class="va">combined_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">purpose</span> <span class="op">==</span> <span class="st">"fullRNASeq"</span>, 
         <span class="op">!</span><span class="va">genotype1</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">grant_df</span><span class="op">$</span><span class="va">GENOTYPE1</span>,
         <span class="va">perturbation1</span> <span class="op">==</span> <span class="st">'deletion'</span>,
         <span class="op">!</span><span class="va">fastqFileName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">grant_induction_set</span>,
         <span class="va">medium</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DMEM"</span><span class="op">)</span>,
         <span class="va">temperature</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">37</span><span class="op">)</span>,
         <span class="va">atmosphere</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CO2"</span><span class="op">)</span>,
         <span class="va">timePoint</span> <span class="op">==</span> <span class="fl">90</span>,
         <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">fastqFileName</span><span class="op">)</span><span class="op">)</span>

<span class="va">other_regulators_set</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="va">induction_set_wt</span>, <span class="va">regulators_not_in_2016_grant</span><span class="op">)</span></code></pre></div>
<p>Extracting overexpressions is similar</p>
</div>
<div id="environmental-perturbation" class="section level1">
<h1 class="hasAnchor">
<a href="#environmental-perturbation" class="anchor"></a>Environmental Perturbation</h1>
<p>The definition of the Environmental Perturbation set is:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">createEnvPertSet</span>
<span class="co">#&gt; function (combined_df) </span>
<span class="co">#&gt; {</span>
<span class="co">#&gt;     combined_df %&gt;% filter((treatment == "" | treatment == "cAMP" | </span>
<span class="co">#&gt;         treatment == "noTreatment" | is.na(treatment)), (experimentDesign == </span>
<span class="co">#&gt;         "Environmental_Perturbation" | experimentDesign == "ep_cAMP_titration"), </span>
<span class="co">#&gt;         purpose == "fullRNASeq", !is.na(fastqFileName), genotype1 == </span>
<span class="co">#&gt;             "CNAG_00000") %&gt;% mutate(timePoint = as.integer(timePoint)) %&gt;% </span>
<span class="co">#&gt;         mutate(across(c("treatment", "otherConditions", "medium", </span>
<span class="co">#&gt;             "atmosphere", "temperature", "timePoint"), ~ifelse(. == </span>
<span class="co">#&gt;             "", NA, .))) %&gt;% mutate(treatmentConc = replace_na(treatmentConc, </span>
<span class="co">#&gt;         "noTreatmentConc")) %&gt;% mutate(pH = replace_na(pH, "noPH")) %&gt;% </span>
<span class="co">#&gt;         mutate(treatment = replace_na(treatment, "noTreatment")) %&gt;% </span>
<span class="co">#&gt;         mutate(otherConditions = replace_na(otherConditions, </span>
<span class="co">#&gt;             "noOtherConditions")) %&gt;% mutate(medium = replace_na(medium, </span>
<span class="co">#&gt;         "noMedium")) %&gt;% mutate(atmosphere = replace_na(atmosphere, </span>
<span class="co">#&gt;         "noAtmosphere")) %&gt;% mutate(temperature = replace_na(temperature, </span>
<span class="co">#&gt;         "noTemperature")) %&gt;% mutate(timePoint = replace_na(timePoint, </span>
<span class="co">#&gt;         "noTimepoint")) %&gt;% mutate(fastqFileName = str_remove(fastqFileName, </span>
<span class="co">#&gt;         ".fastq.gz")) %&gt;% rename_with(toupper)</span>
<span class="co">#&gt; }</span>
<span class="co">#&gt; &lt;bytecode: 0x55a48ad396f8&gt;</span>
<span class="co">#&gt; &lt;environment: namespace:brentlabRnaSeqTools&gt;</span></code></pre></div>
</div>
<div id="quality-assessment-1-filter" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-assessment-1-filter" class="anchor"></a>Quality Assessment 1 Filter</h1>
<p>To filter <strong>out</strong> all samples which fail auto audit or manual audit (but retaining those which are manual pass regardless of auto audit status), use the following function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qual_assess_1_passing</span> <span class="op">=</span> <span class="fu"><a href="../reference/qualityAssessmentFilter.html">qualityAssessmentFilter</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
