<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>qcExplorer • brentlabRnaSeqTools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="qcExplorer">
<meta property="og:description" content="brentlabRnaSeqTools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">brentlabRnaSeqTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/create_nextflow_samplesheet.html">create_nextflow_samplesheet</a>
    </li>
    <li>
      <a href="../articles/create_set.html">create_set</a>
    </li>
    <li>
      <a href="../articles/database_interaction.html">database_interaction</a>
    </li>
    <li>
      <a href="../articles/experiment_sets.html">experiment_sets</a>
    </li>
    <li>
      <a href="../articles/qcExplorer.html">qcExplorer</a>
    </li>
    <li>
      <a href="../articles/running_the_nextflow_pipeline.html">running_the_nextflow_pipeline</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="qcExplorer_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>qcExplorer</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/brentlabRnaSeqTools/blob/master/vignettes/qcExplorer.Rmd"><code>vignettes/qcExplorer.Rmd</code></a></small>
      <div class="hidden name"><code>qcExplorer.Rmd</code></div>

    </div>

    
    
<div id="under-development" class="section level1">
<h1 class="hasAnchor">
<a href="#under-development" class="anchor"></a>UNDER DEVELOPMENT</h1>
<p>These functions are under active development. Admittedly, there is not much difference between this set of functions and the rest in terms of how ‘under development’ they are. But, these functions rely on the libraries <a href="https://mastering-shiny.org/">shiny</a> and its derivative, <a href="https://rstudio.github.io/shinydashboard/">shinydashboard</a>, and I am still learning this framework. As a result, there is a good deal of restructuring that will be happening as I learn more about how best to structure a shiny app, especially one inside of another package like this.</p>
<p>That said, in functionality, not much will change. Mostly what you will see is changes in the names of arguments and their documentation – right now, the language in the documentation is a bit rough. More placeholders and ‘notes to self’ than effective documentation for others, though I do expect that if you are interested in using these functions that with some clarifying questions the documentation will be sufficient, especially with the examples below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cmatKhan/brentlabRnaSeqTools">brentlabRnaSeqTools</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="creating-the-qc-database" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-the-qc-database" class="anchor"></a>Creating the QC Database</h1>
<p>The main input to <code><a href="../reference/qcExplorer.html">qcExplorer()</a></code> is a SQLite database. To create this database, we need the following:</p>
<ol style="list-style-type: decimal">
<li>A “sample_df”, or metadata *</li>
<li>A directory of tables, in tsv format with the extension .txt *</li>
<li>A qc metric df, sometimes called status_df in the function documentation. This name will evolve into something more descriptive eventually.</li>
</ol>
<ul>
<li>Items 1 and 2 must share 1 column, the id column. This column must uniquely identify the samples</li>
</ul>
<div id="compiling-the-qualimaps-output" class="section level2">
<h2 class="hasAnchor">
<a href="#compiling-the-qualimaps-output" class="anchor"></a>Compiling the Qualimaps output</h2>
<p>This is not compiled by multiqc (part of the nf-co/rnaseq pipeline). You may find the qualimap output in the star_[salmon/RSEM] subdirectory of the nf-co/rnaseq results directory.</p>
<p>Here is an example of how to compile the qualimap output into a qc sheet – note, this is only for genomic origin information.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pool_plate1_qualimaps_files</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html">Sys.glob</a></span><span class="op">(</span><span class="st">"/mnt/htcf_lts/llfs_rnaseq/nf_co_rnaseq_pipeline_results/pool_plate1_results/star_salmon/qualimap/*/rnaseq_qc_results.txt"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pool_plate1_qualimaps_files</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span><span class="op">(</span><span class="va">pool_plate1_qualimaps_files</span><span class="op">)</span><span class="op">)</span>

<span class="va">pool_plate2_qualimaps_files</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html">Sys.glob</a></span><span class="op">(</span><span class="st">"/mnt/htcf_lts/llfs_rnaseq/nf_co_rnaseq_pipeline_results/pool_plate2_results/star_salmon/qualimap/*/rnaseq_qc_results.txt"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pool_plate2_qualimaps_files</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span><span class="op">(</span><span class="va">pool_plate2_qualimaps_files</span><span class="op">)</span><span class="op">)</span>

<span class="va">pool_plate1_qualimap_df</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pool_plate1_qualimaps_files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/compileQualimapSummaryMetrics.html">compileQualimapSummaryMetrics</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">pool_plate1_qualimaps_files</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="fu">write_tsv</span><span class="op">(</span><span class="va">pool_plate1_qualimap_df</span>, <span class="st">"data/pool_plate1_multiqc/qualimap_genomicOrigins.txt"</span><span class="op">)</span>

<span class="va">pool_plate2_qualimap_df</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pool_plate2_qualimaps_files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/compileQualimapSummaryMetrics.html">compileQualimapSummaryMetrics</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">pool_plate2_qualimaps_files</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="fu">write_tsv</span><span class="op">(</span><span class="va">pool_plate2_qualimap_df</span>, <span class="st">"data/pool_plate2_multiqc/qualimap_genomicOrigins.txt"</span><span class="op">)</span>

<span class="va">qualimap_df</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">pool_plate1_qualimap_df</span>, <span class="va">pool_plate2_qualimap_df</span><span class="op">)</span>

<span class="fu">write_csv</span><span class="op">(</span><span class="va">qualimap_df</span>, <span class="st">"data/pool_plate1qualimap_both_pools.csv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="compile-the-tin-py-output" class="section level2">
<h2 class="hasAnchor">
<a href="#compile-the-tin-py-output" class="anchor"></a>Compile the tin.py output</h2>
<p>If you ran tin.py (part of RseQC), then you may want to use the following function (part of the brentlabRnaSeqTools) package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">tin_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/compileTinOutput.html">compileTinOutput</a></span><span class="op">(</span><span class="st">"data/tin_output"</span><span class="op">)</span>

<span class="va">tin_df_augment</span> <span class="op">=</span> <span class="va">tin_df</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>Sample <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">Sample</span>, <span class="st">".genome_accepted_hits.bam"</span><span class="op">)</span><span class="op">)</span>
  

<span class="fu">write_csv</span><span class="op">(</span><span class="va">tiny_df_augment</span>, <span class="st">"data/compiled_tin_both_pools.csv"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="sample_df-and-status_df" class="section level1">
<h1 class="hasAnchor">
<a href="#sample_df-and-status_df" class="anchor"></a>sample_df and status_df</h1>
<div id="sample-df" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-df" class="anchor"></a>sample df</h2>
<p>sample_df is sample metadata – you need to provide that. Critically, it must have a common column with all QC tables. It should also include any batch parameters you wish to explore</p>
</div>
<div id="status-df" class="section level2">
<h2 class="hasAnchor">
<a href="#status-df" class="anchor"></a>status df</h2>
<p>status_df may be created manually, or with the function below. Feel free to add to it as needed. It must include all metrics that you wish to use in the qcExplorer</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">status_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/createStatusDf.html">createStatusDf</a></span><span class="op">(</span><span class="st">""</span>, <span class="st">"data/pool_plate1_multiqc"</span><span class="op">)</span></code></pre></div>
<p>In this case, we want to combine multiple batches of QC tables before making the database</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pool_plate1_paths</span> <span class="op">=</span> <span class="fu"><a href="../reference/getNfCoMultiqcDataPaths.html">getNfCoMultiqcDataPaths</a></span><span class="op">(</span><span class="st">"data/pool_plate1_multiqc/"</span><span class="op">)</span>
<span class="va">pool_plate2_paths</span> <span class="op">=</span> <span class="fu"><a href="../reference/getNfCoMultiqcDataPaths.html">getNfCoMultiqcDataPaths</a></span><span class="op">(</span><span class="st">"data/pool_plate2_multiqc/"</span><span class="op">)</span>

<span class="va">pool_plate1_df_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">pool_plate1_paths</span>, <span class="va">read_tsv</span>, name_repair <span class="op">=</span> <span class="st">'minimal'</span><span class="op">)</span>

<span class="va">pool_plate2_df_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">pool_plate2_paths</span>, <span class="va">read_tsv</span>, name_repair <span class="op">=</span> <span class="st">'minimal'</span><span class="op">)</span>

<span class="va">all_qc_tables</span> <span class="op">=</span> <span class="fu">map2</span><span class="op">(</span><span class="va">pool_plate1_df_list</span>, <span class="va">pool_plate2_df_list</span>, <span class="va">bind_rows</span><span class="op">)</span>

<span class="co"># removing na rows b/c this dataframe also stores lane info when present</span>
<span class="va">all_qc_tables</span><span class="op">$</span><span class="va">multiqc_general_stats</span> <span class="op">=</span> <span class="va">all_qc_tables</span><span class="op">$</span><span class="va">multiqc_general_stats</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">`Samtools_mqc-generalstats-samtools-flagstat_total`</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">database_path</span> <span class="op">=</span> <span class="st">"data/qc_database.sqlite"</span>
<span class="va">id_field</span> <span class="op">=</span> <span class="st">"Sample"</span>

<span class="fu"><a href="../reference/createDatabase.html">createDatabase</a></span><span class="op">(</span><span class="va">sample_df</span>, <span class="va">status_df</span>, <span class="va">all_qc_tables</span>, <span class="va">database_path</span>, <span class="va">id_field</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">batch_choices</span> <span class="op">=</span> <span class="st">"batch"</span>

<span class="fu"><a href="../reference/qcExplorer.html">qcExplorer</a></span><span class="op">(</span><span class="va">database_path</span>, <span class="st">"sample"</span>, <span class="st">"qc_metrics"</span>, <span class="va">id_field</span>, <span class="va">batch_choices</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
